// @spartan-ng/ui-accordion-brain/esm2022/index.mjs
import { NgModule } from '@angular/core';
import { BrnAccordionContentComponent } from './lib/brn-accordion-content.component';
import { BrnAccordionItemDirective } from './lib/brn-accordion-item.directive';
import { BrnAccordionTriggerDirective } from './lib/brn-accordion-trigger.directive';
import { BrnAccordionDirective } from './lib/brn-accordion.directive';
import * as i0 from "@angular/core";
export * from './lib/brn-accordion-content.component';
export * from './lib/brn-accordion-item.directive';
export * from './lib/brn-accordion-trigger.directive';
export * from './lib/brn-accordion.directive';
export const BrnAccordionImports = [
    BrnAccordionDirective,
    BrnAccordionContentComponent,
    BrnAccordionItemDirective,
    BrnAccordionTriggerDirective,
];
export class BrnAccordionModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnAccordionModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.5", ngImport: i0, type: BrnAccordionModule, imports: [BrnAccordionDirective,
            BrnAccordionContentComponent,
            BrnAccordionItemDirective,
            BrnAccordionTriggerDirective], exports: [BrnAccordionDirective,
            BrnAccordionContentComponent,
            BrnAccordionItemDirective,
            BrnAccordionTriggerDirective] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnAccordionModule }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnAccordionModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...BrnAccordionImports],
                    exports: [...BrnAccordionImports],
                }]
        }] });
