// @spartan-ng/ui-hovercard-brain/esm2022/index.mjs
import { NgModule } from '@angular/core';
import { BrnHoverCardContentDirective } from './lib/brn-hover-card-content.directive';
import { BrnHoverCardTriggerDirective } from './lib/brn-hover-card-trigger.directive';
import { BrnHoverCardComponent } from './lib/brn-hover-card.component';
import * as i0 from "@angular/core";
export { BrnHoverCardContentDirective } from './lib/brn-hover-card-content.directive';
export { BrnHoverCardContentService } from './lib/brn-hover-card-content.service';
export { BrnHoverCardTriggerDirective } from './lib/brn-hover-card-trigger.directive';
export { BrnHoverCardComponent } from './lib/brn-hover-card.component';
export { createHoverObservable } from './lib/createHoverObservable';
export const BrnHoverCardImports = [
    BrnHoverCardComponent,
    BrnHoverCardContentDirective,
    BrnHoverCardTriggerDirective,
];
export class BrnHoverCardModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnHoverCardModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.5", ngImport: i0, type: BrnHoverCardModule, imports: [BrnHoverCardComponent,
            BrnHoverCardContentDirective,
            BrnHoverCardTriggerDirective], exports: [BrnHoverCardComponent,
            BrnHoverCardContentDirective,
            BrnHoverCardTriggerDirective] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnHoverCardModule }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnHoverCardModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...BrnHoverCardImports],
                    exports: [...BrnHoverCardImports],
                }]
        }] });
