// @spartan-ng/ui-tabs-brain/esm2022/lib/brn-tabs-list.directive.mjs
import { FocusKeyManager } from '@angular/cdk/a11y';
import { ContentChildren, Directive, inject } from '@angular/core';
import { rxHostListener } from '@spartan-ng/ui-core';
import { take } from 'rxjs';
import { BrnTabsTriggerDirective } from './brn-tabs-trigger.directive';
import { BrnTabsDirective } from './brn-tabs.directive';
import * as i0 from "@angular/core";
export class BrnTabsListDirective {
    constructor() {
        this._root = inject(BrnTabsDirective);
        this._orientation = this._root.$orientation;
        this._direction = this._root.$direction;
        this._value = this._root.$value;
        this._tabs = this._root.$tabs;
        this._keyDownListener = rxHostListener('keydown');
    }
    ngAfterContentInit() {
        if (!this.triggers) {
            return;
        }
        this._keyManager = new FocusKeyManager(this.triggers)
            .withHorizontalOrientation(this._direction())
            .withHomeAndEnd()
            .withPageUpDown()
            .withWrap();
        // needed because by default the index is set to -1, which means first interaction is skipped
        this._keyDownListener.pipe(take(1)).subscribe(() => {
            const currentKey = this._value();
            let activeIndex = 0;
            if (currentKey && this.triggers) {
                const currentTab = this._tabs[currentKey];
                if (currentTab) {
                    activeIndex = this.triggers.toArray().indexOf(currentTab.trigger);
                }
            }
            this._keyManager?.setActiveItem(activeIndex);
        });
        this._keyDownListener.subscribe((event) => {
            if ('key' in event) {
                if (this._orientation() === 'horizontal') {
                    if (event.key === 'ArrowUp' || event.key === 'ArrowDown')
                        return;
                }
                if (this._orientation() === 'vertical') {
                    if (event.key === 'ArrowLeft' || event.key === 'ArrowRight')
                        return;
                }
            }
            this._keyManager?.onKeydown(event);
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnTabsListDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.5", type: BrnTabsListDirective, isStandalone: true, selector: "[brnTabsList]", host: { attributes: { "role": "tablist" }, properties: { "attr.aria-orientation": "_orientation()", "attr.data-orientation": "_orientation()" } }, queries: [{ propertyName: "triggers", predicate: BrnTabsTriggerDirective, descendants: true }], exportAs: ["brnTabsList"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnTabsListDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[brnTabsList]',
                    standalone: true,
                    host: {
                        role: 'tablist',
                        '[attr.aria-orientation]': '_orientation()',
                        '[attr.data-orientation]': '_orientation()',
                    },
                    exportAs: 'brnTabsList',
                }]
        }], propDecorators: { triggers: [{
                type: ContentChildren,
                args: [BrnTabsTriggerDirective, { descendants: true }]
            }] } });
