// @spartan-ng/ui-tabs-brain/esm2022/lib/brn-tabs.directive.mjs
import { Directive, EventEmitter, Input, Output, input, signal } from '@angular/core';
import * as i0 from "@angular/core";
export class BrnTabsDirective {
    constructor() {
        this.orientation = input('horizontal');
        /** internal **/
        this.$orientation = this.orientation;
        this.direction = input('ltr');
        /** internal **/
        this.$direction = this.direction;
        this.tabActivated = new EventEmitter();
        // leaving this as an @input and signal to be set programmatically
        // current limitation by InputSignal which are readonly
        this._value = signal(undefined);
        /** internal **/
        this.$value = this._value.asReadonly();
        this.activationMode = input('automatic');
        /** internal **/
        this.$activationMode = this.activationMode;
        this._tabs = {};
        this.$tabs = this._tabs;
    }
    set value(value) {
        this._value.set(value);
    }
    registerTrigger(key, trigger) {
        this._tabs[key] = {
            ...(this._tabs[key] ?? {}),
            trigger,
        };
    }
    registerContent(key, content) {
        this._tabs[key] = {
            ...(this._tabs[key] ?? {}),
            content,
        };
    }
    emitTabActivated(key) {
        this.tabActivated.emit(key);
    }
    setValue(key) {
        this._value.set(key);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnTabsDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "17.3.5", type: BrnTabsDirective, isStandalone: true, selector: "[brnTabs]", inputs: { orientation: { classPropertyName: "orientation", publicName: "orientation", isSignal: true, isRequired: false, transformFunction: null }, direction: { classPropertyName: "direction", publicName: "direction", isSignal: true, isRequired: false, transformFunction: null }, value: { classPropertyName: "value", publicName: "brnTabs", isSignal: false, isRequired: false, transformFunction: null }, activationMode: { classPropertyName: "activationMode", publicName: "activationMode", isSignal: true, isRequired: false, transformFunction: null } }, outputs: { tabActivated: "tabActivated" }, host: { properties: { "attr.data-orientation": "orientation()", "attr.dir": "direction()" } }, exportAs: ["brnTabs"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnTabsDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[brnTabs]',
                    standalone: true,
                    host: {
                        '[attr.data-orientation]': 'orientation()',
                        '[attr.dir]': 'direction()',
                    },
                    exportAs: 'brnTabs',
                }]
        }], propDecorators: { tabActivated: [{
                type: Output
            }], value: [{
                type: Input,
                args: ['brnTabs']
            }] } });
