// @spartan-ng/ui-tabs-brain/esm2022/lib/brn-tabs-trigger.directive.mjs
import { Directive, ElementRef, Input, computed, effect, inject, input } from '@angular/core';
import { BrnTabsDirective } from './brn-tabs.directive';
import * as i0 from "@angular/core";
export class BrnTabsTriggerDirective {
    constructor() {
        this.elementRef = inject(ElementRef);
        this._root = inject(BrnTabsDirective);
        this._orientation = this._root.$orientation;
        this.triggerFor = input.required({ alias: 'brnTabsTrigger' });
        this.selected = computed(() => this._root.$value() === this.triggerFor());
        this.contentId = computed(() => 'brn-tabs-content-' + this.triggerFor());
        this.labelId = computed(() => 'brn-tabs-label-' + this.triggerFor());
        // leaving this as an @input to be compatible with the `FocusKeyManager` used in the `BrnTabsListDirective`
        this.disabled = false;
        effect(() => {
            this._root.registerTrigger(this.triggerFor(), this);
        });
    }
    focus() {
        this.elementRef.nativeElement.focus();
        if (this._root.$activationMode() === 'automatic') {
            this.activate();
        }
    }
    activate() {
        if (!this.triggerFor())
            return;
        this._root.setValue(this.triggerFor());
        this._root.emitTabActivated(this.triggerFor());
    }
    get key() {
        return this.triggerFor();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnTabsTriggerDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "17.3.5", type: BrnTabsTriggerDirective, isStandalone: true, selector: "button[brnTabsTrigger]", inputs: { triggerFor: { classPropertyName: "triggerFor", publicName: "brnTabsTrigger", isSignal: true, isRequired: true, transformFunction: null }, disabled: { classPropertyName: "disabled", publicName: "disabled", isSignal: false, isRequired: false, transformFunction: null } }, host: { attributes: { "type": "button", "role": "tab" }, listeners: { "click": "activate()" }, properties: { "id": "labelId()", "tabindex": "selected() ? \"0\": \"-1\"", "attr.aria-selected": "selected()", "attr.aria-controls": "contentId()", "attr.data-state": "selected() ? 'active' : 'inactive'", "attr.data-orientation": "_orientation()", "attr.data-disabled": "disabled ? '' : undefined" } }, exportAs: ["brnTabsTrigger"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnTabsTriggerDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'button[brnTabsTrigger]',
                    standalone: true,
                    host: {
                        '[id]': 'labelId()',
                        type: 'button',
                        role: 'tab',
                        '[tabindex]': 'selected() ? "0": "-1"',
                        '[attr.aria-selected]': 'selected()',
                        '[attr.aria-controls]': 'contentId()',
                        '[attr.data-state]': "selected() ? 'active' : 'inactive'",
                        '[attr.data-orientation]': '_orientation()',
                        '[attr.data-disabled]': "disabled ? '' : undefined",
                        '(click)': 'activate()',
                    },
                    exportAs: 'brnTabsTrigger',
                }]
        }], ctorParameters: () => [], propDecorators: { disabled: [{
                type: Input
            }] } });
