// @ngxpert/cmdk/esm2022/lib/directives/loader/loader.directive.mjs
import { Directive, TemplateRef, ViewContainerRef, inject, } from '@angular/core';
import * as i0 from "@angular/core";
export class LoaderDirective {
    constructor() {
        this._hasView = false;
        this._templateRef = inject(TemplateRef);
        this._viewContainer = inject(ViewContainerRef);
    }
    set cmdkLoader(condition) {
        if (condition && !this._hasView) {
            this._viewContainer.createEmbeddedView(this._templateRef);
            this._hasView = true;
        }
        else if (!condition && this._hasView) {
            this._viewContainer.clear();
            this._hasView = false;
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: LoaderDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: LoaderDirective, isStandalone: true, selector: "[cmdkLoader]", host: { classAttribute: "cmdk-loader" }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: LoaderDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[cmdkLoader]',
                    // eslint-disable-next-line @angular-eslint/no-host-metadata-property
                    host: {
                        class: 'cmdk-loader',
                    },
                    standalone: true,
                }]
        }] });
