// @spartan-ng/ui-switch-brain/esm2022/index.mjs
import { NgModule } from '@angular/core';
import { BrnSwitchThumbComponent } from './lib/brn-switch-thumb.component';
import { BrnSwitchComponent } from './lib/brn-switch.component';
import * as i0 from "@angular/core";
export * from './lib/brn-switch-thumb.component';
export * from './lib/brn-switch.component';
export const BrnSwitchImports = [BrnSwitchComponent, BrnSwitchThumbComponent];
export class BrnSwitchModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnSwitchModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.5", ngImport: i0, type: BrnSwitchModule, imports: [BrnSwitchComponent, BrnSwitchThumbComponent], exports: [BrnSwitchComponent, BrnSwitchThumbComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnSwitchModule, imports: [BrnSwitchThumbComponent] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnSwitchModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...BrnSwitchImports],
                    exports: [...BrnSwitchImports],
                }]
        }] });
