// @spartan-ng/ui-dialog-brain/esm2022/lib/brn-dialog-content.directive.mjs
import { Directive, Input, TemplateRef, computed, inject } from '@angular/core';
import { provideExposesStateProviderExisting } from '@spartan-ng/ui-core';
import { BrnDialogRef } from './brn-dialog-ref';
import { BrnDialogComponent } from './brn-dialog.component';
import * as i0 from "@angular/core";
export class BrnDialogContentDirective {
    constructor() {
        this._brnDialog = inject(BrnDialogComponent, { optional: true });
        this._brnDialogRef = inject(BrnDialogRef, { optional: true });
        this._template = inject(TemplateRef);
        this.state = computed(() => this._brnDialog?.state() ?? this._brnDialogRef?.state() ?? 'closed');
        if (!this._brnDialog)
            return;
        this._brnDialog.registerTemplate(this._template);
    }
    set class(newClass) {
        if (!this._brnDialog)
            return;
        this._brnDialog.setPanelClass(newClass);
    }
    set context(context) {
        if (!this._brnDialog)
            return;
        this._brnDialog.setContext(context);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnDialogContentDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.5", type: BrnDialogContentDirective, isStandalone: true, selector: "[brnDialogContent]", inputs: { class: "class", context: "context" }, providers: [provideExposesStateProviderExisting(() => BrnDialogContentDirective)], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnDialogContentDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[brnDialogContent]',
                    standalone: true,
                    providers: [provideExposesStateProviderExisting(() => BrnDialogContentDirective)],
                }]
        }], ctorParameters: () => [], propDecorators: { class: [{
                type: Input
            }], context: [{
                type: Input
            }] } });
