// @spartan-ng/ui-dialog-brain/esm2022/lib/brn-dialog-trigger.directive.mjs
import { Directive, Input, inject, input, signal } from '@angular/core';
import { BrnDialogRef } from './brn-dialog-ref';
import { BrnDialogComponent } from './brn-dialog.component';
import * as i0 from "@angular/core";
let idSequence = 0;
export class BrnDialogTriggerDirective {
    constructor() {
        this._brnDialog = inject(BrnDialogComponent, { optional: true });
        this._brnDialogRef = inject(BrnDialogRef, { optional: true });
        this.id = input(`brn-dialog-trigger-${idSequence++}`);
        this.state = this._brnDialogRef?.state ?? signal('closed');
        this.dialogId = `brn-dialog-${this._brnDialogRef?.dialogId ?? idSequence++}`;
    }
    set brnDialogTriggerFor(brnDialog) {
        this._brnDialog = brnDialog;
    }
    open() {
        this._brnDialog?.open();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnDialogTriggerDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "17.3.5", type: BrnDialogTriggerDirective, isStandalone: true, selector: "button[brnDialogTrigger],button[brnDialogTriggerFor]", inputs: { id: { classPropertyName: "id", publicName: "id", isSignal: true, isRequired: false, transformFunction: null }, brnDialogTriggerFor: { classPropertyName: "brnDialogTriggerFor", publicName: "brnDialogTriggerFor", isSignal: false, isRequired: false, transformFunction: null } }, host: { attributes: { "aria-haspopup": "dialog" }, listeners: { "click": "open()" }, properties: { "id": "id()", "attr.aria-expanded": "state() === 'open' ? 'true': 'false'", "attr.data-state": "state()", "attr.aria-controls": "dialogId" } }, exportAs: ["brnDialogTrigger"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnDialogTriggerDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'button[brnDialogTrigger],button[brnDialogTriggerFor]',
                    standalone: true,
                    host: {
                        '[id]': 'id()',
                        '(click)': 'open()',
                        'aria-haspopup': 'dialog',
                        '[attr.aria-expanded]': "state() === 'open' ? 'true': 'false'",
                        '[attr.data-state]': 'state()',
                        '[attr.aria-controls]': 'dialogId',
                    },
                    exportAs: 'brnDialogTrigger',
                }]
        }], propDecorators: { brnDialogTriggerFor: [{
                type: Input
            }] } });
