// @spartan-ng/ui-dialog-brain/esm2022/lib/brn-dialog-title.directive.mjs
import { Directive, effect, inject, signal } from '@angular/core';
import { BrnDialogRef } from './brn-dialog-ref';
import * as i0 from "@angular/core";
export class BrnDialogTitleDirective {
    constructor() {
        this._brnDialogRef = inject(BrnDialogRef);
        this._id = signal(`brn-dialog-title-${this._brnDialogRef?.dialogId}`);
        effect(() => {
            this._brnDialogRef.setAriaLabelledBy(this._id());
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnDialogTitleDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.5", type: BrnDialogTitleDirective, isStandalone: true, selector: "[brnDialogTitle]", host: { properties: { "id": "_id()" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnDialogTitleDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[brnDialogTitle]',
                    standalone: true,
                    host: {
                        '[id]': '_id()',
                    },
                }]
        }], ctorParameters: () => [] });
