// @spartan-ng/ui-dialog-brain/esm2022/lib/brn-dialog-overlay.component.mjs
import { ChangeDetectionStrategy, Component, Input, ViewEncapsulation, inject } from '@angular/core';
import { provideCustomClassSettableExisting } from '@spartan-ng/ui-core';
import { BrnDialogComponent } from './brn-dialog.component';
import * as i0 from "@angular/core";
export class BrnDialogOverlayComponent {
    constructor() {
        this._brnDialog = inject(BrnDialogComponent);
    }
    set class(newClass) {
        this._brnDialog.setOverlayClass(newClass);
    }
    setClassToCustomElement(newClass) {
        this._brnDialog.setOverlayClass(newClass);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnDialogOverlayComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.3.5", type: BrnDialogOverlayComponent, isStandalone: true, selector: "brn-dialog-overlay", inputs: { class: "class" }, providers: [provideCustomClassSettableExisting(() => BrnDialogOverlayComponent)], ngImport: i0, template: ``, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnDialogOverlayComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'brn-dialog-overlay',
                    standalone: true,
                    template: ``,
                    providers: [provideCustomClassSettableExisting(() => BrnDialogOverlayComponent)],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    encapsulation: ViewEncapsulation.None,
                }]
        }], propDecorators: { class: [{
                type: Input
            }] } });
