// @spartan-ng/ui-dialog-brain/esm2022/lib/brn-dialog-ref.mjs
import { Subject, take } from 'rxjs';
import { cssClassesToArray } from './brn-dialog.service';
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export class BrnDialogRef {
    get open() {
        return this.state() === 'open';
    }
    constructor(_cdkDialogRef, _open, state, dialogId, _options) {
        this._cdkDialogRef = _cdkDialogRef;
        this._open = _open;
        this.state = state;
        this.dialogId = dialogId;
        this._options = _options;
        this._closing$ = new Subject();
        this.closing$ = this._closing$.asObservable();
        this.closed$ = this._cdkDialogRef.closed.pipe(take(1));
    }
    close(result, delay = this._options?.closeDelay ?? 0) {
        if (!this.open || this._options?.disableClose)
            return;
        this._closing$.next();
        this._open.set(false);
        if (this._previousTimeout) {
            clearTimeout(this._previousTimeout);
        }
        this._previousTimeout = setTimeout(() => {
            this._cdkDialogRef.close(result);
        }, delay);
    }
    setPanelClass(paneClass) {
        this._cdkDialogRef.config.panelClass = cssClassesToArray(paneClass);
    }
    setOverlayClass(overlayClass) {
        this._cdkDialogRef.config.backdropClass = cssClassesToArray(overlayClass);
    }
    setAriaDescribedBy(ariaDescribedBy) {
        this._cdkDialogRef.config.ariaDescribedBy = ariaDescribedBy;
    }
    setAriaLabelledBy(ariaLabelledBy) {
        this._cdkDialogRef.config.ariaLabelledBy = ariaLabelledBy;
    }
    setAriaLabel(ariaLabel) {
        this._cdkDialogRef.config.ariaLabel = ariaLabel;
    }
}
