// @spartan-ng/ui-dialog-brain/esm2022/lib/brn-dialog-close.directive.mjs
import { coerceNumberProperty } from '@angular/cdk/coercion';
import { Directive, Input, inject } from '@angular/core';
import { BrnDialogRef } from './brn-dialog-ref';
import * as i0 from "@angular/core";
export class BrnDialogCloseDirective {
    constructor() {
        this._brnDialogRef = inject(BrnDialogRef);
    }
    set delay(value) {
        this._delay = coerceNumberProperty(value);
    }
    close() {
        this._brnDialogRef.close(undefined, this._delay);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnDialogCloseDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.5", type: BrnDialogCloseDirective, isStandalone: true, selector: "button[brnDialogClose]", inputs: { delay: "delay" }, host: { listeners: { "click": "close()" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnDialogCloseDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'button[brnDialogClose]',
                    standalone: true,
                    host: {
                        '(click)': 'close()',
                    },
                }]
        }], propDecorators: { delay: [{
                type: Input
            }] } });
