// @spartan-ng/ui-select-brain/esm2022/lib/brn-select-trigger.directive.mjs
import { Directive, ElementRef, computed, inject } from '@angular/core';
import { NgControl } from '@angular/forms';
import { BrnSelectService } from './brn-select.service';
import * as i0 from "@angular/core";
export class BrnSelectTriggerDirective {
    constructor() {
        this.el = inject(ElementRef);
        this._selectService = inject(BrnSelectService);
        this._ngControl = inject(NgControl, { optional: true });
        this.isExpanded = this._selectService.isExpanded;
        this.selectTriggerId = computed(() => `${this._selectService.id()}--trigger`);
        this.selectContentId = computed(() => `${this._selectService.id()}--content`);
        this.selectDisable = computed(() => this._selectService.disabled());
        this.selectTriggerLabelledBy = computed(() => {
            if (this._selectService.value() && this._selectService.value().length > 0) {
                return `${this._selectService.labelId()} ${this._selectService.id()}--value`;
            }
            return this._selectService.labelId();
        });
        if (!this._selectService)
            return;
        this._selectService._setSelectTrigger(this);
    }
    ngAfterViewInit() {
        this._selectService.setTriggerWidth(this.el.nativeElement.offsetWidth);
    }
    focus() {
        this.el.nativeElement.focus();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnSelectTriggerDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.5", type: BrnSelectTriggerDirective, isStandalone: true, selector: "[brnSelectTrigger]", host: { attributes: { "role": "combobox", "aria-autocomplete": "none", "type": "button" }, properties: { "attr.id": "selectTriggerId()", "disabled": "selectDisable()", "attr.aria-expanded": "isExpanded()", "attr.aria-controls": "selectContentId() + ''", "attr.aria-labelledBy": "selectTriggerLabelledBy()", "attr.dir": "_selectService.dir()", "class.ng-invalid": "this._ngControl?.invalid || null", "class.ng-dirty": "this._ngControl?.dirty || null", "class.ng-valid": "this._ngControl?.valid || null", "class.ng-touched": "this._ngControl?.touched || null", "class.ng-untouched": "this._ngControl?.untouched || null", "class.ng-pristine": "this._ngControl?.pristine || null" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnSelectTriggerDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[brnSelectTrigger]',
                    standalone: true,
                    host: {
                        role: 'combobox',
                        '[attr.id]': 'selectTriggerId()',
                        '[disabled]': 'selectDisable()',
                        '[attr.aria-expanded]': 'isExpanded()',
                        '[attr.aria-controls]': "selectContentId() + ''",
                        '[attr.aria-labelledBy]': 'selectTriggerLabelledBy()',
                        'aria-autocomplete': 'none',
                        '[attr.dir]': '_selectService.dir()',
                        '[class.ng-invalid]': 'this._ngControl?.invalid || null',
                        '[class.ng-dirty]': 'this._ngControl?.dirty || null',
                        '[class.ng-valid]': 'this._ngControl?.valid || null',
                        '[class.ng-touched]': 'this._ngControl?.touched || null',
                        '[class.ng-untouched]': 'this._ngControl?.untouched || null',
                        '[class.ng-pristine]': 'this._ngControl?.pristine || null',
                        type: 'button',
                    },
                }]
        }], ctorParameters: () => [] });
