// @spartan-ng/ui-select-brain/esm2022/lib/brn-select-label.directive.mjs
import { Directive, inject } from '@angular/core';
import { BrnLabelDirective } from '@spartan-ng/ui-label-brain';
import { BrnSelectGroupDirective } from './brn-select-group.directive';
import * as i0 from "@angular/core";
import * as i1 from "@spartan-ng/ui-label-brain";
export class BrnSelectLabelDirective {
    constructor() {
        this.group = inject(BrnSelectGroupDirective, { optional: true });
        this.label = inject(BrnLabelDirective, { host: true });
        this.group?.labelledBy.set(this.label.id);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnSelectLabelDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.5", type: BrnSelectLabelDirective, isStandalone: true, selector: "[brnSelectLabel]", hostDirectives: [{ directive: i1.BrnLabelDirective }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnSelectLabelDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[brnSelectLabel]',
                    hostDirectives: [BrnLabelDirective],
                    standalone: true,
                }]
        }], ctorParameters: () => [] });
