// @spartan-ng/ui-select-brain/esm2022/lib/brn-select-option.directive.mjs
import { CdkOption } from '@angular/cdk/listbox';
import { Directive, ElementRef, Input, computed, inject, signal } from '@angular/core';
import { takeUntilDestroyed, toObservable } from '@angular/core/rxjs-interop';
import { BrnSelectService } from './brn-select.service';
import * as i0 from "@angular/core";
import * as i1 from "@angular/cdk/listbox";
export class BrnSelectOptionDirective {
    constructor() {
        this._cdkSelectOption = inject(CdkOption, { host: true });
        this._selectService = inject(BrnSelectService);
        this._selected = signal(false);
        this._focused = signal(false);
        this.elementRef = inject(ElementRef);
        this.selected = computed(() => this._selected());
        this.focused = computed(() => this._focused());
        this.checkedState = computed(() => (this._selected() ? 'checked' : 'unchecked'));
        this.dir = computed(() => this._selectService.dir());
        this._disabled = false;
        this._selectService.registerOption(this._cdkSelectOption);
        toObservable(this._selectService.value)
            .pipe(takeUntilDestroyed())
            .subscribe((selectedValues) => {
            if (Array.isArray(selectedValues)) {
                const itemFound = selectedValues.find((val) => val === this._cdkSelectOption.value);
                this._selected.set(!!itemFound);
            }
            else {
                this._selected.set(this._cdkSelectOption.value === selectedValues);
            }
        });
    }
    ngOnDestroy() {
        this._selectService.deregisterOption(this._cdkSelectOption);
    }
    set value(value) {
        this._cdkSelectOption.value = value;
    }
    set disabled(value) {
        this._disabled = value;
    }
    get disabled() {
        return this._disabled;
    }
    hover() {
        this.focus();
    }
    focus() {
        this._cdkSelectOption.focus();
        this._focused.set(true);
    }
    blur() {
        this._focused.set(false);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnSelectOptionDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.5", type: BrnSelectOptionDirective, isStandalone: true, selector: "[brnOption]", inputs: { value: "value", disabled: "disabled" }, host: { listeners: { "mouseenter": "hover()", "blur": "blur()" }, properties: { "attr.dir": "_selectService.dir()" } }, hostDirectives: [{ directive: i1.CdkOption }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnSelectOptionDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[brnOption]',
                    standalone: true,
                    hostDirectives: [CdkOption],
                    host: {
                        '(mouseenter)': 'hover()',
                        '(blur)': 'blur()',
                        '[attr.dir]': '_selectService.dir()',
                    },
                }]
        }], ctorParameters: () => [], propDecorators: { value: [{
                type: Input
            }], disabled: [{
                type: Input
            }] } });
