// @spartan-ng/ui-select-brain/esm2022/lib/brn-select-value.component.mjs
import { ChangeDetectionStrategy, Component, Input, computed, inject } from '@angular/core';
import { BrnSelectService } from './brn-select.service';
import * as i0 from "@angular/core";
export class BrnSelectValueComponent {
    constructor() {
        this._selectService = inject(BrnSelectService);
        this.id = computed(() => `${this._selectService.id()}--value`);
        this.placeholder = computed(() => this._selectService.placeholder());
        this.value = computed(() => {
            const value = this._selectService.selectedOptions();
            if (value.length === 0) {
                return null;
            }
            const selectedLabels = value.map((selectedOption) => selectedOption?.getLabel());
            if (this._selectService.dir() === 'rtl') {
                selectedLabels.reverse();
            }
            return this.transformFn(selectedLabels);
        });
        this.transformFn = (values) => (values ?? []).join(', ');
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnSelectValueComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.3.5", type: BrnSelectValueComponent, isStandalone: true, selector: "brn-select-value, hlm-select-value", inputs: { transformFn: "transformFn" }, host: { properties: { "id": "id()" } }, ngImport: i0, template: `
		{{ value() || placeholder() }}
	`, isInline: true, styles: [":host{display:-webkit-box;-webkit-box-orient:vertical;-webkit-line-clamp:1;white-space:nowrap;pointer-events:none}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnSelectValueComponent, decorators: [{
            type: Component,
            args: [{ selector: 'brn-select-value, hlm-select-value', template: `
		{{ value() || placeholder() }}
	`, host: {
                        '[id]': 'id()',
                    }, standalone: true, changeDetection: ChangeDetectionStrategy.OnPush, styles: [":host{display:-webkit-box;-webkit-box-orient:vertical;-webkit-line-clamp:1;white-space:nowrap;pointer-events:none}\n"] }]
        }], propDecorators: { 
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        transformFn: [{
                type: Input
            }] } });
