// @spartan-ng/ui-select-brain/esm2022/lib/brn-select-group.directive.mjs
import { Directive, signal } from '@angular/core';
import * as i0 from "@angular/core";
export class BrnSelectGroupDirective {
    constructor() {
        this.labelledBy = signal('');
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnSelectGroupDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.5", type: BrnSelectGroupDirective, isStandalone: true, selector: "[brnSelectGroup]", host: { attributes: { "role": "group" }, properties: { "attr.aria-labelledby": "labelledBy()" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnSelectGroupDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[brnSelectGroup]',
                    standalone: true,
                    host: {
                        role: 'group',
                        '[attr.aria-labelledby]': 'labelledBy()',
                    },
                }]
        }] });
