// @spartan-ng/ui-label-brain/esm2022/lib/brn-label.directive.mjs
import { isPlatformBrowser } from '@angular/common';
import { Directive, ElementRef, Input, PLATFORM_ID, inject, signal } from '@angular/core';
import { NgControl } from '@angular/forms';
import * as i0 from "@angular/core";
let nextId = 0;
export class BrnLabelDirective {
    constructor() {
        this._id = signal(`brn-label-${nextId++}`);
        this._ngControl = inject(NgControl, { optional: true });
        this._isBrowser = isPlatformBrowser(inject(PLATFORM_ID));
        this._element = inject(ElementRef).nativeElement;
        this._dataDisabled = signal('auto');
        this.dataDisabled = this._dataDisabled.asReadonly();
    }
    set id(id) {
        this._id.set(id || this._id());
    }
    get id() {
        return this._id();
    }
    ngOnInit() {
        if (!this._isBrowser)
            return;
        this._changes = new MutationObserver((mutations) => {
            mutations.forEach((mutation) => {
                if (mutation.attributeName !== 'data-disabled')
                    return;
                // eslint-disable-next-line
                const state = mutation.target.attributes.getNamedItem(mutation.attributeName)?.value === 'true';
                this._dataDisabled.set(state ?? 'auto');
            });
        });
        this._changes?.observe(this._element, {
            attributes: true,
            childList: true,
            characterData: true,
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnLabelDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.5", type: BrnLabelDirective, isStandalone: true, selector: "[brnLabel]", inputs: { id: "id" }, host: { properties: { "id": "_id()", "class.ng-invalid": "this._ngControl?.invalid || null", "class.ng-dirty": "this._ngControl?.dirty || null", "class.ng-valid": "this._ngControl?.valid || null", "class.ng-touched": "this._ngControl?.touched || null" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnLabelDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[brnLabel]',
                    standalone: true,
                    host: {
                        '[id]': '_id()',
                        '[class.ng-invalid]': 'this._ngControl?.invalid || null',
                        '[class.ng-dirty]': 'this._ngControl?.dirty || null',
                        '[class.ng-valid]': 'this._ngControl?.valid || null',
                        '[class.ng-touched]': 'this._ngControl?.touched || null',
                    },
                }]
        }], propDecorators: { id: [{
                type: Input
            }] } });
