// @spartan-ng/ui-hovercard-brain/esm2022/lib/brn-hover-card.component.mjs
import { Component, ContentChild } from '@angular/core';
import { BrnHoverCardContentDirective } from './brn-hover-card-content.directive';
import { BrnHoverCardContentService } from './brn-hover-card-content.service';
import { BrnHoverCardTriggerDirective } from './brn-hover-card-trigger.directive';
import * as i0 from "@angular/core";
export class BrnHoverCardComponent {
    ngAfterContentInit() {
        if (!this._trigger || !this._content)
            return;
        this._trigger.brnHoverCardTriggerFor = this._content;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnHoverCardComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.3.5", type: BrnHoverCardComponent, isStandalone: true, selector: "brn-hover-card", providers: [BrnHoverCardContentService], queries: [{ propertyName: "_trigger", first: true, predicate: BrnHoverCardTriggerDirective, descendants: true, static: true }, { propertyName: "_content", first: true, predicate: BrnHoverCardContentDirective, descendants: true, static: true }], ngImport: i0, template: `
		<ng-content />
	`, isInline: true }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnHoverCardComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'brn-hover-card',
                    standalone: true,
                    providers: [BrnHoverCardContentService],
                    template: `
		<ng-content />
	`,
                }]
        }], propDecorators: { _trigger: [{
                type: ContentChild,
                args: [BrnHoverCardTriggerDirective, { static: true }]
            }], _content: [{
                type: ContentChild,
                args: [BrnHoverCardContentDirective, { static: true }]
            }] } });
