// @spartan-ng/ui-hovercard-brain/esm2022/lib/brn-hover-card-content.directive.mjs
import { Directive, TemplateRef, inject } from '@angular/core';
import { provideExposedSideProviderExisting, provideExposesStateProviderExisting, } from '@spartan-ng/ui-core';
import { BrnHoverCardContentService } from './brn-hover-card-content.service';
import * as i0 from "@angular/core";
export class BrnHoverCardContentDirective {
    constructor() {
        this._contentService = inject(BrnHoverCardContentService);
        this.state = this._contentService.state;
        this.side = this._contentService.side;
        this.template = inject(TemplateRef);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnHoverCardContentDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.5", type: BrnHoverCardContentDirective, isStandalone: true, selector: "[brnHoverCardContent]", providers: [
            provideExposedSideProviderExisting(() => BrnHoverCardContentDirective),
            provideExposesStateProviderExisting(() => BrnHoverCardContentDirective),
        ], exportAs: ["brnHoverCardContent"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnHoverCardContentDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[brnHoverCardContent]',
                    standalone: true,
                    exportAs: 'brnHoverCardContent',
                    providers: [
                        provideExposedSideProviderExisting(() => BrnHoverCardContentDirective),
                        provideExposesStateProviderExisting(() => BrnHoverCardContentDirective),
                    ],
                }]
        }] });
