// @spartan-ng/ui-select-brain/esm2022/index.mjs
import { NgModule } from '@angular/core';
import { BrnSelectContentComponent } from './lib/brn-select-content.component';
import { BrnSelectGroupDirective } from './lib/brn-select-group.directive';
import { BrnSelectLabelDirective } from './lib/brn-select-label.directive';
import { BrnSelectOptionDirective } from './lib/brn-select-option.directive';
import { BrnSelectScrollDownDirective } from './lib/brn-select-scroll-down.directive';
import { BrnSelectScrollUpDirective } from './lib/brn-select-scroll-up.directive';
import { BrnSelectTriggerDirective } from './lib/brn-select-trigger.directive';
import { BrnSelectValueComponent } from './lib/brn-select-value.component';
import { BrnSelectComponent } from './lib/brn-select.component';
import * as i0 from "@angular/core";
export * from './lib/brn-select-content.component';
export * from './lib/brn-select-group.directive';
export * from './lib/brn-select-label.directive';
export * from './lib/brn-select-option.directive';
export * from './lib/brn-select-scroll-down.directive';
export * from './lib/brn-select-scroll-up.directive';
export * from './lib/brn-select-trigger.directive';
export * from './lib/brn-select-value.component';
export * from './lib/brn-select.component';
export * from './lib/brn-select.service';
export const BrnSelectImports = [
    BrnSelectComponent,
    BrnSelectContentComponent,
    BrnSelectTriggerDirective,
    BrnSelectOptionDirective,
    BrnSelectValueComponent,
    BrnSelectScrollDownDirective,
    BrnSelectScrollUpDirective,
    BrnSelectGroupDirective,
    BrnSelectLabelDirective,
];
export class BrnSelectModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnSelectModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.5", ngImport: i0, type: BrnSelectModule, imports: [BrnSelectComponent,
            BrnSelectContentComponent,
            BrnSelectTriggerDirective,
            BrnSelectOptionDirective,
            BrnSelectValueComponent,
            BrnSelectScrollDownDirective,
            BrnSelectScrollUpDirective,
            BrnSelectGroupDirective,
            BrnSelectLabelDirective], exports: [BrnSelectComponent,
            BrnSelectContentComponent,
            BrnSelectTriggerDirective,
            BrnSelectOptionDirective,
            BrnSelectValueComponent,
            BrnSelectScrollDownDirective,
            BrnSelectScrollUpDirective,
            BrnSelectGroupDirective,
            BrnSelectLabelDirective] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnSelectModule, imports: [BrnSelectComponent] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnSelectModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...BrnSelectImports],
                    exports: [...BrnSelectImports],
                }]
        }] });
