// @spartan-ng/ui-menu-brain/esm2022/index.mjs
import { NgModule } from '@angular/core';
import { BrnContextMenuTriggerDirective } from './lib/brn-context-menu-trigger.directive';
import { BrnMenuBarDirective } from './lib/brn-menu-bar.directive';
import { BrnMenuGroupDirective } from './lib/brn-menu-group.directive';
import { BrnMenuItemCheckboxDirective } from './lib/brn-menu-item-checkbox.directive';
import { BrnMenuItemRadioDirective } from './lib/brn-menu-item-radio.directive';
import { BrnMenuItemDirective } from './lib/brn-menu-item.directive';
import { BrnMenuTriggerDirective } from './lib/brn-menu-trigger.directive';
import { BrnMenuDirective } from './lib/brn-menu.directive';
import * as i0 from "@angular/core";
export * from './lib/brn-context-menu-trigger.directive';
export * from './lib/brn-menu-bar.directive';
export * from './lib/brn-menu-group.directive';
export * from './lib/brn-menu-item-checkbox.directive';
export * from './lib/brn-menu-item-radio.directive';
export * from './lib/brn-menu-item.directive';
export * from './lib/brn-menu-trigger.directive';
export * from './lib/brn-menu.directive';
export const BrnMenuItemImports = [
    BrnMenuGroupDirective,
    BrnMenuItemDirective,
    BrnMenuItemRadioDirective,
    BrnMenuItemCheckboxDirective,
];
export const BrnMenuImports = [BrnMenuTriggerDirective, BrnMenuDirective, ...BrnMenuItemImports];
export const BrnMenuBarImports = [...BrnMenuImports, BrnMenuBarDirective];
export const BrnContextMenuImports = [...BrnMenuImports, BrnContextMenuTriggerDirective];
export class BrnMenuItemModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnMenuItemModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.5", ngImport: i0, type: BrnMenuItemModule, imports: [BrnMenuGroupDirective,
            BrnMenuItemDirective,
            BrnMenuItemRadioDirective,
            BrnMenuItemCheckboxDirective], exports: [BrnMenuGroupDirective,
            BrnMenuItemDirective,
            BrnMenuItemRadioDirective,
            BrnMenuItemCheckboxDirective] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnMenuItemModule }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnMenuItemModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...BrnMenuItemImports],
                    exports: [...BrnMenuItemImports],
                }]
        }] });
export class BrnMenuModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnMenuModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.5", ngImport: i0, type: BrnMenuModule, imports: [BrnMenuTriggerDirective, BrnMenuDirective, BrnMenuGroupDirective,
            BrnMenuItemDirective,
            BrnMenuItemRadioDirective,
            BrnMenuItemCheckboxDirective], exports: [BrnMenuTriggerDirective, BrnMenuDirective, BrnMenuGroupDirective,
            BrnMenuItemDirective,
            BrnMenuItemRadioDirective,
            BrnMenuItemCheckboxDirective] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnMenuModule }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnMenuModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...BrnMenuImports],
                    exports: [...BrnMenuImports],
                }]
        }] });
export class BrnMenuBarModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnMenuBarModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.5", ngImport: i0, type: BrnMenuBarModule, imports: [BrnMenuTriggerDirective, BrnMenuDirective, BrnMenuGroupDirective,
            BrnMenuItemDirective,
            BrnMenuItemRadioDirective,
            BrnMenuItemCheckboxDirective, BrnMenuBarDirective], exports: [BrnMenuTriggerDirective, BrnMenuDirective, BrnMenuGroupDirective,
            BrnMenuItemDirective,
            BrnMenuItemRadioDirective,
            BrnMenuItemCheckboxDirective, BrnMenuBarDirective] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnMenuBarModule }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnMenuBarModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...BrnMenuBarImports],
                    exports: [...BrnMenuBarImports],
                }]
        }] });
export class BrnContextMenuModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnContextMenuModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.5", ngImport: i0, type: BrnContextMenuModule, imports: [BrnMenuTriggerDirective, BrnMenuDirective, BrnMenuGroupDirective,
            BrnMenuItemDirective,
            BrnMenuItemRadioDirective,
            BrnMenuItemCheckboxDirective, BrnContextMenuTriggerDirective], exports: [BrnMenuTriggerDirective, BrnMenuDirective, BrnMenuGroupDirective,
            BrnMenuItemDirective,
            BrnMenuItemRadioDirective,
            BrnMenuItemCheckboxDirective, BrnContextMenuTriggerDirective] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnContextMenuModule }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnContextMenuModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...BrnContextMenuImports],
                    exports: [...BrnContextMenuImports],
                }]
        }] });
