// @ngxpert/cmdk/esm2022/lib/cmdk.module.mjs
import { NgModule } from '@angular/core';
import { InputDirective } from './directives/input/input.directive';
import { EmptyDirective } from './directives/empty/empty.directive';
import { CommandComponent } from './components/command/command.component';
import { DynamicViewModule } from '@ngneat/overview';
import { GroupComponent } from './components/group/group.component';
import { SeparatorComponent } from './components/separator/separator.component';
import { CommonModule } from '@angular/common';
import { ItemDirective } from './directives/item/item.directive';
import { A11yModule } from '@angular/cdk/a11y';
import { LoaderDirective } from './directives/loader/loader.directive';
import { ListComponent } from './components/list/list.component';
import * as i0 from "@angular/core";
const ComponentsAndDirectives = [
    CommandComponent,
    InputDirective,
    EmptyDirective,
    GroupComponent,
    SeparatorComponent,
    ItemDirective,
    LoaderDirective,
    ListComponent,
];
export class CmdkModule {
    static forRoot() {
        return {
            ngModule: CmdkModule,
        };
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: CmdkModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: CmdkModule, imports: [CommonModule, DynamicViewModule, A11yModule, CommandComponent,
            InputDirective,
            EmptyDirective,
            GroupComponent,
            SeparatorComponent,
            ItemDirective,
            LoaderDirective,
            ListComponent], exports: [CommandComponent,
            InputDirective,
            EmptyDirective,
            GroupComponent,
            SeparatorComponent,
            ItemDirective,
            LoaderDirective,
            ListComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: CmdkModule, imports: [CommonModule, DynamicViewModule, A11yModule, GroupComponent] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: CmdkModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, DynamicViewModule, A11yModule, ...ComponentsAndDirectives],
                    exports: ComponentsAndDirectives,
                }]
        }] });
