// @spartan-ng/ui-menu-brain/esm2022/lib/brn-menu-item.directive.mjs
import { CdkMenuItem } from '@angular/cdk/menu';
import { Directive, Input, Output, booleanAttribute, inject, signal } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/cdk/menu";
export class BrnMenuItemDirective {
    constructor() {
        this._cdkMenuItem = inject(CdkMenuItem, { host: true });
        this._disabled = signal(this._cdkMenuItem.disabled);
        this.triggered = this._cdkMenuItem.triggered;
    }
    set disabled(value) {
        this._cdkMenuItem.disabled = value;
        this._disabled.set(value);
    }
    get disabled() {
        return this._disabled();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnMenuItemDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "16.1.0", version: "17.3.5", type: BrnMenuItemDirective, isStandalone: true, selector: "[brnMenuItem]", inputs: { disabled: ["disabled", "disabled", booleanAttribute] }, outputs: { triggered: "triggered" }, host: { properties: { "disabled": "_disabled()" } }, hostDirectives: [{ directive: i1.CdkMenuItem }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnMenuItemDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[brnMenuItem]',
                    standalone: true,
                    hostDirectives: [CdkMenuItem],
                    host: {
                        '[disabled]': '_disabled()',
                    },
                }]
        }], propDecorators: { disabled: [{
                type: Input,
                args: [{ transform: booleanAttribute }]
            }], triggered: [{
                type: Output
            }] } });
