// @spartan-ng/ui-menu-brain/esm2022/lib/brn-menu-item-checkbox.directive.mjs
import { CdkMenuItemCheckbox } from '@angular/cdk/menu';
import { Directive, Input, Output, booleanAttribute, inject, signal } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/cdk/menu";
export class BrnMenuItemCheckboxDirective {
    constructor() {
        this._cdkMenuItem = inject(CdkMenuItemCheckbox, { host: true });
        this._checked = signal(this._cdkMenuItem.checked);
        this._disabled = signal(this._cdkMenuItem.disabled);
        this.triggered = this._cdkMenuItem.triggered;
    }
    set checked(value) {
        this._cdkMenuItem.checked = value;
        this._checked.set(value);
    }
    get checked() {
        return this._checked();
    }
    set disabled(value) {
        this._cdkMenuItem.disabled = value;
        this._disabled.set(value);
    }
    get disabled() {
        return this._disabled();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnMenuItemCheckboxDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "16.1.0", version: "17.3.5", type: BrnMenuItemCheckboxDirective, isStandalone: true, selector: "[brnMenuItemCheckbox]", inputs: { checked: ["checked", "checked", booleanAttribute], disabled: ["disabled", "disabled", booleanAttribute] }, outputs: { triggered: "triggered" }, host: { properties: { "class.checked": "_checked()", "disabled": "_disabled()" } }, hostDirectives: [{ directive: i1.CdkMenuItemCheckbox }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnMenuItemCheckboxDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[brnMenuItemCheckbox]',
                    standalone: true,
                    hostDirectives: [CdkMenuItemCheckbox],
                    host: {
                        '[class.checked]': '_checked()',
                        '[disabled]': '_disabled()',
                    },
                }]
        }], propDecorators: { checked: [{
                type: Input,
                args: [{ transform: booleanAttribute }]
            }], disabled: [{
                type: Input,
                args: [{ transform: booleanAttribute }]
            }], triggered: [{
                type: Output
            }] } });
