// @spartan-ng/ui-menu-brain/esm2022/lib/brn-context-menu-trigger.directive.mjs
import { CdkContextMenuTrigger } from '@angular/cdk/menu';
import { Directive, Input, effect, inject, signal } from '@angular/core';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import * as i0 from "@angular/core";
import * as i1 from "@angular/cdk/menu";
export class BrnContextMenuTriggerDirective {
    set align(value) {
        this._align.set(value);
    }
    set brnCtxMenuTriggerFor(value) {
        this._cdkTrigger.menuTemplateRef = value;
    }
    constructor() {
        this._cdkTrigger = inject(CdkContextMenuTrigger, { host: true });
        this._align = signal(undefined);
        // once the trigger opens we wait until the next tick and then grab the last position
        // used to position the menu. we store this in our trigger which the brnMenu directive has
        // access to through DI
        this._cdkTrigger.opened.pipe(takeUntilDestroyed()).subscribe(() => setTimeout(() => 
        // eslint-disable-next-line
        (this._cdkTrigger['_spartanLastPosition'] = // eslint-disable-next-line
            this._cdkTrigger.overlayRef._positionStrategy._lastPosition)));
        effect(() => {
            const align = this._align();
            if (!align)
                return;
            this._cdkTrigger.menuPosition = [
                {
                    originX: align,
                    originY: 'bottom',
                    overlayX: align,
                    overlayY: 'top',
                },
                {
                    originX: align,
                    originY: 'top',
                    overlayX: align,
                    overlayY: 'bottom',
                },
            ];
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnContextMenuTriggerDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.5", type: BrnContextMenuTriggerDirective, isStandalone: true, selector: "[brnCtxMenuTriggerFor]", inputs: { align: "align", brnCtxMenuTriggerFor: "brnCtxMenuTriggerFor" }, hostDirectives: [{ directive: i1.CdkContextMenuTrigger }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnContextMenuTriggerDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[brnCtxMenuTriggerFor]',
                    standalone: true,
                    hostDirectives: [CdkContextMenuTrigger],
                }]
        }], ctorParameters: () => [], propDecorators: { align: [{
                type: Input
            }], brnCtxMenuTriggerFor: [{
                type: Input
            }] } });
