// @spartan-ng/ui-avatar-brain/esm2022/index.mjs
import { NgModule } from '@angular/core';
import { BrnAvatarComponent } from './lib/brn-avatar.component';
import { BrnAvatarFallbackDirective } from './lib/fallback';
import { BrnAvatarImageDirective } from './lib/image';
import * as i0 from "@angular/core";
export * from './lib/brn-avatar.component';
export * from './lib/fallback';
export * from './lib/image';
export * from './lib/util';
export const BrnAvatarImports = [BrnAvatarComponent, BrnAvatarFallbackDirective, BrnAvatarImageDirective];
export class BrnAvatarModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnAvatarModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.5", ngImport: i0, type: BrnAvatarModule, imports: [BrnAvatarComponent, BrnAvatarFallbackDirective, BrnAvatarImageDirective], exports: [BrnAvatarComponent, BrnAvatarFallbackDirective, BrnAvatarImageDirective] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnAvatarModule }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnAvatarModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...BrnAvatarImports],
                    exports: [...BrnAvatarImports],
                }]
        }] });
