// @spartan-ng/ui-tooltip-brain/esm2022/index.mjs
import { NgModule } from '@angular/core';
import { BrnTooltipContentComponent } from './lib/brn-tooltip-content.component';
import { BrnTooltipContentDirective } from './lib/brn-tooltip-content.directive';
import { BrnTooltipTriggerDirective } from './lib/brn-tooltip-trigger.directive';
import { BrnTooltipDirective } from './lib/brn-tooltip.directive';
import * as i0 from "@angular/core";
export * from './lib/brn-tooltip-content.component';
export * from './lib/brn-tooltip-content.directive';
export * from './lib/brn-tooltip-trigger.directive';
export * from './lib/brn-tooltip.directive';
export const BrnTooltipImports = [
    BrnTooltipDirective,
    BrnTooltipContentDirective,
    BrnTooltipTriggerDirective,
    BrnTooltipContentComponent,
];
export class BrnTooltipModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnTooltipModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.5", ngImport: i0, type: BrnTooltipModule, imports: [BrnTooltipDirective,
            BrnTooltipContentDirective,
            BrnTooltipTriggerDirective,
            BrnTooltipContentComponent], exports: [BrnTooltipDirective,
            BrnTooltipContentDirective,
            BrnTooltipTriggerDirective,
            BrnTooltipContentComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnTooltipModule }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnTooltipModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...BrnTooltipImports],
                    exports: [...BrnTooltipImports],
                }]
        }] });
