// @spartan-ng/ui-alertdialog-brain/esm2022/index.mjs
import { NgModule } from '@angular/core';
import { BrnAlertDialogContentDirective } from './lib/brn-alert-dialog-content.directive';
import { BrnAlertDialogDescriptionDirective } from './lib/brn-alert-dialog-description.directive';
import { BrnAlertDialogOverlayComponent } from './lib/brn-alert-dialog-overlay.component';
import { BrnAlertDialogTitleDirective } from './lib/brn-alert-dialog-title.directive';
import { BrnAlertDialogTriggerDirective } from './lib/brn-alert-dialog-trigger.directive';
import { BrnAlertDialogComponent } from './lib/brn-alert-dialog.component';
import * as i0 from "@angular/core";
export * from './lib/brn-alert-dialog-content.directive';
export * from './lib/brn-alert-dialog-description.directive';
export * from './lib/brn-alert-dialog-overlay.component';
export * from './lib/brn-alert-dialog-title.directive';
export * from './lib/brn-alert-dialog-trigger.directive';
export * from './lib/brn-alert-dialog.component';
export const BrnAlertDialogImports = [
    BrnAlertDialogComponent,
    BrnAlertDialogOverlayComponent,
    BrnAlertDialogTriggerDirective,
    BrnAlertDialogContentDirective,
    BrnAlertDialogTitleDirective,
    BrnAlertDialogDescriptionDirective,
];
export class BrnAlertDialogModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnAlertDialogModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.5", ngImport: i0, type: BrnAlertDialogModule, imports: [BrnAlertDialogComponent,
            BrnAlertDialogOverlayComponent,
            BrnAlertDialogTriggerDirective,
            BrnAlertDialogContentDirective,
            BrnAlertDialogTitleDirective,
            BrnAlertDialogDescriptionDirective], exports: [BrnAlertDialogComponent,
            BrnAlertDialogOverlayComponent,
            BrnAlertDialogTriggerDirective,
            BrnAlertDialogContentDirective,
            BrnAlertDialogTitleDirective,
            BrnAlertDialogDescriptionDirective] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnAlertDialogModule }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnAlertDialogModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...BrnAlertDialogImports],
                    exports: [...BrnAlertDialogImports],
                }]
        }] });
