// @spartan-ng/ui-avatar-brain/esm2022/lib/image/brn-avatar-image.directive.mjs
import { Directive, HostListener, computed, signal } from '@angular/core';
import * as i0 from "@angular/core";
export class BrnAvatarImageDirective {
    constructor() {
        this.error = signal(false);
        this.loaded = signal(false);
        this.canShow = computed(() => {
            return this.loaded() && !this.error();
        });
    }
    onError() {
        this.error.set(true);
    }
    onLoad() {
        this.loaded.set(true);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnAvatarImageDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.5", type: BrnAvatarImageDirective, isStandalone: true, selector: "img[brnAvatarImage]", host: { listeners: { "error": "onError()", "load": "onLoad()" } }, exportAs: ["avatarImage"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnAvatarImageDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'img[brnAvatarImage]',
                    standalone: true,
                    exportAs: 'avatarImage',
                }]
        }], propDecorators: { onError: [{
                type: HostListener,
                args: ['error']
            }], onLoad: [{
                type: HostListener,
                args: ['load']
            }] } });
