// @spartan-ng/ui-dialog-brain/esm2022/index.mjs
import { NgModule } from '@angular/core';
import { BrnDialogCloseDirective } from './lib/brn-dialog-close.directive';
import { BrnDialogContentDirective } from './lib/brn-dialog-content.directive';
import { BrnDialogDescriptionDirective } from './lib/brn-dialog-description.directive';
import { BrnDialogOverlayComponent } from './lib/brn-dialog-overlay.component';
import { BrnDialogTitleDirective } from './lib/brn-dialog-title.directive';
import { BrnDialogTriggerDirective } from './lib/brn-dialog-trigger.directive';
import { BrnDialogComponent } from './lib/brn-dialog.component';
import * as i0 from "@angular/core";
export * from './lib/brn-dialog-close.directive';
export * from './lib/brn-dialog-content.directive';
export * from './lib/brn-dialog-description.directive';
export * from './lib/brn-dialog-options';
export * from './lib/brn-dialog-overlay.component';
export * from './lib/brn-dialog-ref';
export * from './lib/brn-dialog-state';
export * from './lib/brn-dialog-title.directive';
export * from './lib/brn-dialog-trigger.directive';
export * from './lib/brn-dialog.component';
export * from './lib/brn-dialog.service';
export const BrnDialogImports = [
    BrnDialogComponent,
    BrnDialogOverlayComponent,
    BrnDialogTriggerDirective,
    BrnDialogCloseDirective,
    BrnDialogContentDirective,
    BrnDialogTitleDirective,
    BrnDialogDescriptionDirective,
];
export class BrnDialogModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnDialogModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.5", ngImport: i0, type: BrnDialogModule, imports: [BrnDialogComponent,
            BrnDialogOverlayComponent,
            BrnDialogTriggerDirective,
            BrnDialogCloseDirective,
            BrnDialogContentDirective,
            BrnDialogTitleDirective,
            BrnDialogDescriptionDirective], exports: [BrnDialogComponent,
            BrnDialogOverlayComponent,
            BrnDialogTriggerDirective,
            BrnDialogCloseDirective,
            BrnDialogContentDirective,
            BrnDialogTitleDirective,
            BrnDialogDescriptionDirective] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnDialogModule }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnDialogModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...BrnDialogImports],
                    exports: [...BrnDialogImports],
                }]
        }] });
