// @spartan-ng/ui-popover-brain/esm2022/index.mjs
import { NgModule } from '@angular/core';
import { BrnPopoverCloseDirective } from './lib/brn-popover-close.directive';
import { BrnPopoverContentDirective } from './lib/brn-popover-content.directive';
import { BrnPopoverTriggerDirective } from './lib/brn-popover-trigger.directive';
import { BrnPopoverComponent } from './lib/brn-popover.component';
import * as i0 from "@angular/core";
export * from './lib/brn-popover-close.directive';
export * from './lib/brn-popover-content.directive';
export * from './lib/brn-popover-trigger.directive';
export * from './lib/brn-popover.component';
export const BrnPopoverImports = [
    BrnPopoverComponent,
    BrnPopoverTriggerDirective,
    BrnPopoverCloseDirective,
    BrnPopoverContentDirective,
];
export class BrnPopoverModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnPopoverModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.5", ngImport: i0, type: BrnPopoverModule, imports: [BrnPopoverComponent,
            BrnPopoverTriggerDirective,
            BrnPopoverCloseDirective,
            BrnPopoverContentDirective], exports: [BrnPopoverComponent,
            BrnPopoverTriggerDirective,
            BrnPopoverCloseDirective,
            BrnPopoverContentDirective] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnPopoverModule }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnPopoverModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...BrnPopoverImports],
                    exports: [...BrnPopoverImports],
                }]
        }] });
