// @spartan-ng/ui-separator-brain/fesm2022/spartan-ng-ui-separator-brain.mjs
import * as i0 from '@angular/core';
import { booleanAttribute, Component, Input, NgModule } from '@angular/core';

class BrnSeparatorComponent {
    constructor() {
        this._orientation = 'horizontal';
        this._decorative = false;
    }
    set orientation(value) {
        this._orientation = value;
    }
    set decorative(value) {
        this._decorative = value;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnSeparatorComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "16.1.0", version: "17.3.5", type: BrnSeparatorComponent, isStandalone: true, selector: "brn-separator", inputs: { orientation: "orientation", decorative: ["decorative", "decorative", booleanAttribute] }, host: { properties: { "role": "_decorative ? \"none\" : \"separator\"", "attr.aria-orientation": "_decorative ? undefined : _orientation === \"vertical\" ? \"vertical\" : undefined ", "attr.data-orientation": "_orientation" } }, ngImport: i0, template: '', isInline: true }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnSeparatorComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'brn-separator',
                    standalone: true,
                    template: '',
                    host: {
                        '[role]': '_decorative ? "none" : "separator"',
                        '[attr.aria-orientation]': '_decorative ? undefined : _orientation === "vertical" ? "vertical" : undefined ',
                        '[attr.data-orientation]': '_orientation',
                    },
                }]
        }], propDecorators: { orientation: [{
                type: Input
            }], decorative: [{
                type: Input,
                args: [{ transform: booleanAttribute }]
            }] } });

class BrnSeparatorModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnSeparatorModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.5", ngImport: i0, type: BrnSeparatorModule, imports: [BrnSeparatorComponent], exports: [BrnSeparatorComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnSeparatorModule }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnSeparatorModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [BrnSeparatorComponent],
                    exports: [BrnSeparatorComponent],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { BrnSeparatorComponent, BrnSeparatorModule };
