// @spartan-ng/ui-popover-brain/esm2022/lib/brn-popover.component.mjs
import { ChangeDetectionStrategy, Component, Input, ViewEncapsulation, effect, forwardRef, numberAttribute, signal, untracked, } from '@angular/core';
import { BrnDialogComponent } from '@spartan-ng/ui-dialog-brain';
import * as i0 from "@angular/core";
export class BrnPopoverComponent extends BrnDialogComponent {
    set sideOffset(value) {
        this._sideOffset.set(value);
    }
    set align(value) {
        this._align.set(value);
    }
    constructor() {
        super();
        this._sideOffset = signal(0);
        this._align = signal('center');
        this.hasBackdrop = false;
        this.ariaDescribedBy = '';
        this.ariaLabelledBy = '';
        this.scrollStrategy = this.ssos.reposition();
        effect(() => {
            const align = this._align();
            this.attachPositions = [
                {
                    originX: align,
                    originY: 'bottom',
                    overlayX: align,
                    overlayY: 'top',
                },
                {
                    originX: align,
                    originY: 'top',
                    overlayX: align,
                    overlayY: 'bottom',
                },
            ];
            this.applySideOffset(untracked(this._sideOffset));
        });
        effect(() => {
            this.applySideOffset(this._sideOffset());
        });
    }
    applySideOffset(sideOffset) {
        this.attachPositions = (this._options['attachPositions'] ?? []).map((position) => ({
            ...position,
            offsetY: position.originY === 'top' ? -sideOffset : sideOffset,
        }));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnPopoverComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "16.1.0", version: "17.3.5", type: BrnPopoverComponent, isStandalone: true, selector: "brn-popover", inputs: { sideOffset: ["sideOffset", "sideOffset", numberAttribute], align: "align" }, providers: [
            {
                provide: BrnDialogComponent,
                useExisting: forwardRef(() => BrnPopoverComponent),
            },
        ], exportAs: ["brnPopover"], usesInheritance: true, ngImport: i0, template: `
		<ng-content />
	`, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnPopoverComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'brn-popover',
                    standalone: true,
                    template: `
		<ng-content />
	`,
                    providers: [
                        {
                            provide: BrnDialogComponent,
                            useExisting: forwardRef(() => BrnPopoverComponent),
                        },
                    ],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    encapsulation: ViewEncapsulation.None,
                    exportAs: 'brnPopover',
                }]
        }], ctorParameters: () => [], propDecorators: { sideOffset: [{
                type: Input,
                args: [{ transform: numberAttribute }]
            }], align: [{
                type: Input
            }] } });
