// @spartan-ng/ui-popover-brain/esm2022/lib/brn-popover-content.directive.mjs
import { Directive } from '@angular/core';
import { provideExposesStateProviderExisting } from '@spartan-ng/ui-core';
import { BrnDialogContentDirective } from '@spartan-ng/ui-dialog-brain';
import * as i0 from "@angular/core";
export class BrnPopoverContentDirective extends BrnDialogContentDirective {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnPopoverContentDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.5", type: BrnPopoverContentDirective, isStandalone: true, selector: "[brnPopoverContent]", providers: [provideExposesStateProviderExisting(() => BrnPopoverContentDirective)], usesInheritance: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnPopoverContentDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[brnPopoverContent]',
                    standalone: true,
                    providers: [provideExposesStateProviderExisting(() => BrnPopoverContentDirective)],
                }]
        }] });
