// @spartan-ng/ui-toggle-brain/esm2022/lib/brn-toggle.directive.mjs
import { ChangeDetectorRef, Directive, EventEmitter, Input, Output, booleanAttribute, computed, inject, signal, } from '@angular/core';
import { ToggleGroupCanBeNullableProvider } from './toggle-group-can-be-nullable-provider';
import * as i0 from "@angular/core";
let uniqueId = 0;
export class BrnToggleDirective {
    constructor() {
        this._cdr = inject(ChangeDetectorRef);
        this._tgCanBeNullableProvider = inject(ToggleGroupCanBeNullableProvider, { optional: true });
        this._disabled = signal(undefined);
        this._state = signal('off', { equal: (a, b) => a === b });
        this.state = this._state.asReadonly();
        this.toggleDisabled = this._disabled.asReadonly();
        this.isOn = computed(() => this.state() === 'on');
        this.id = 'brn-toggle-' + uniqueId++;
        this._disableToggleClick = false;
        this.toggled = new EventEmitter();
    }
    set disabled(value) {
        this._disabled.set(value ? true : undefined);
    }
    // eslint-disable-next-line @angular-eslint/no-input-rename
    set setState(value) {
        this._state.set(value);
    }
    set disableToggleClick(value) {
        this._disableToggleClick = value;
    }
    toggle() {
        if (this._disableToggleClick)
            return;
        if (this._state() === 'on') {
            this.toggleOff();
        }
        else {
            this.toggleOn();
        }
    }
    toggleOff() {
        if (this._tgCanBeNullableProvider && !this._tgCanBeNullableProvider._canBeNullable(this.value))
            return;
        this._state.set('off');
        this.toggled.emit('off');
    }
    toggleOn() {
        this._state.set('on');
        this.toggled.emit('on');
    }
    _markForCheck() {
        this._cdr.markForCheck();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnToggleDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "16.1.0", version: "17.3.5", type: BrnToggleDirective, isStandalone: true, selector: "button[hlmToggle], button[brnToggle]", inputs: { id: "id", value: "value", disabled: ["disabled", "disabled", booleanAttribute], setState: ["state", "setState"], disableToggleClick: ["disableToggleClick", "disableToggleClick", booleanAttribute] }, outputs: { toggled: "toggled" }, host: { listeners: { "click": "toggle()" }, properties: { "attr.disabled": "toggleDisabled()", "attr.data-disabled": "toggleDisabled()", "attr.data-state": "state()", "attr.aria-pressed": "isOn()" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnToggleDirective, decorators: [{
            type: Directive,
            args: [{
                    // eslint-disable-next-line @angular-eslint/directive-selector
                    selector: 'button[hlmToggle], button[brnToggle]',
                    standalone: true,
                    host: {
                        '[attr.disabled]': 'toggleDisabled()',
                        '[attr.data-disabled]': 'toggleDisabled()',
                        '[attr.data-state]': 'state()',
                        '[attr.aria-pressed]': 'isOn()',
                        '(click)': 'toggle()',
                    },
                }]
        }], propDecorators: { id: [{
                type: Input
            }], value: [{
                type: Input
            }], disabled: [{
                type: Input,
                args: [{ transform: booleanAttribute }]
            }], setState: [{
                type: Input,
                args: ['state']
            }], disableToggleClick: [{
                type: Input,
                args: [{ transform: booleanAttribute }]
            }], toggled: [{
                type: Output
            }] } });
