// @spartan-ng/ui-alertdialog-brain/esm2022/lib/brn-alert-dialog-overlay.component.mjs
import { ChangeDetectionStrategy, Component, ViewEncapsulation } from '@angular/core';
import { provideCustomClassSettableExisting } from '@spartan-ng/ui-core';
import { BrnDialogOverlayComponent } from '@spartan-ng/ui-dialog-brain';
import * as i0 from "@angular/core";
export class BrnAlertDialogOverlayComponent extends BrnDialogOverlayComponent {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnAlertDialogOverlayComponent, deps: null, target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.3.5", type: BrnAlertDialogOverlayComponent, isStandalone: true, selector: "brn-alert-dialog-overlay", providers: [provideCustomClassSettableExisting(() => BrnAlertDialogOverlayComponent)], usesInheritance: true, ngImport: i0, template: ``, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnAlertDialogOverlayComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'brn-alert-dialog-overlay',
                    standalone: true,
                    providers: [provideCustomClassSettableExisting(() => BrnAlertDialogOverlayComponent)],
                    template: ``,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    encapsulation: ViewEncapsulation.None,
                }]
        }] });
