// @spartan-ng/ui-alertdialog-brain/esm2022/lib/brn-alert-dialog-content.directive.mjs
import { Directive } from '@angular/core';
import { provideExposesStateProviderExisting } from '@spartan-ng/ui-core';
import { BrnDialogContentDirective } from '@spartan-ng/ui-dialog-brain';
import * as i0 from "@angular/core";
export class BrnAlertDialogContentDirective extends BrnDialogContentDirective {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnAlertDialogContentDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.5", type: BrnAlertDialogContentDirective, isStandalone: true, selector: "[brnAlertDialogContent]", providers: [provideExposesStateProviderExisting(() => BrnAlertDialogContentDirective)], usesInheritance: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnAlertDialogContentDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[brnAlertDialogContent]',
                    standalone: true,
                    providers: [provideExposesStateProviderExisting(() => BrnAlertDialogContentDirective)],
                }]
        }] });
