// @spartan-ng/ui-alertdialog-brain/esm2022/lib/brn-alert-dialog.component.mjs
import { ChangeDetectionStrategy, Component, ViewEncapsulation, forwardRef } from '@angular/core';
import { BrnDialogComponent } from '@spartan-ng/ui-dialog-brain';
import * as i0 from "@angular/core";
export class BrnAlertDialogComponent extends BrnDialogComponent {
    constructor() {
        super();
        this._options.role = 'alertdialog';
        this._options.closeOnBackdropClick = false;
        this._options.closeOnOutsidePointerEvents = false;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnAlertDialogComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.3.5", type: BrnAlertDialogComponent, isStandalone: true, selector: "brn-alert-dialog", providers: [
            {
                provide: BrnDialogComponent,
                useExisting: forwardRef(() => BrnAlertDialogComponent),
            },
        ], exportAs: ["brnAlertDialog"], usesInheritance: true, ngImport: i0, template: `
		<ng-content />
	`, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnAlertDialogComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'brn-alert-dialog',
                    standalone: true,
                    template: `
		<ng-content />
	`,
                    providers: [
                        {
                            provide: BrnDialogComponent,
                            useExisting: forwardRef(() => BrnAlertDialogComponent),
                        },
                    ],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    encapsulation: ViewEncapsulation.None,
                    exportAs: 'brnAlertDialog',
                }]
        }], ctorParameters: () => [] });
