// @spartan-ng/ui-progress-brain/esm2022/lib/brn-progress-indicator.component.mjs
import { Component, inject } from '@angular/core';
import { BrnProgressComponent } from './brn-progress.component';
import * as i0 from "@angular/core";
export class BrnProgressIndicatorComponent {
    constructor() {
        this._parent = inject(BrnProgressComponent);
        this.progressState = this._parent.progressState;
        this.max = this._parent.$max;
        this.value = this._parent.$value;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnProgressIndicatorComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.3.5", type: BrnProgressIndicatorComponent, isStandalone: true, selector: "brn-progress-indicator", host: { properties: { "attr.data-state": "progressState()", "attr.data-value": "value() ?? undefined", "attr.data-max": "max()" } }, ngImport: i0, template: '', isInline: true }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnProgressIndicatorComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'brn-progress-indicator',
                    standalone: true,
                    template: '',
                    host: {
                        '[attr.data-state]': 'progressState()',
                        '[attr.data-value]': 'value() ?? undefined',
                        '[attr.data-max]': 'max()',
                    },
                }]
        }] });
