// @spartan-ng/ui-collapsible-brain/esm2022/lib/brn-collapsible.component.mjs
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, Input, ViewEncapsulation, booleanAttribute, inject, signal, } from '@angular/core';
import * as i0 from "@angular/core";
let collapsibleContentIdSequence = 0;
export class BrnCollapsibleComponent {
    constructor() {
        this._cdr = inject(ChangeDetectorRef);
        this.state = signal('closed');
        this.contentId = signal('brn-collapsible-content-' + collapsibleContentIdSequence++);
        this._disabled = signal(undefined);
        this.collapsibleDisabled = this._disabled.asReadonly();
    }
    set disabled(value) {
        this._disabled.set(value ? true : undefined);
    }
    toggle() {
        this.state.set(this.state() === 'closed' ? 'open' : 'closed');
        this._cdr.detectChanges();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnCollapsibleComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "16.1.0", version: "17.3.5", type: BrnCollapsibleComponent, isStandalone: true, selector: "brn-collapsible", inputs: { disabled: ["disabled", "disabled", booleanAttribute] }, host: { properties: { "attr.data-state": "state()", "attr.disabled": "collapsibleDisabled()" } }, ngImport: i0, template: `
		<ng-content />
	`, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnCollapsibleComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'brn-collapsible',
                    standalone: true,
                    host: {
                        '[attr.data-state]': 'state()',
                        '[attr.disabled]': 'collapsibleDisabled()',
                    },
                    template: `
		<ng-content />
	`,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    encapsulation: ViewEncapsulation.None,
                }]
        }], propDecorators: { disabled: [{
                type: Input,
                args: [{ transform: booleanAttribute }]
            }] } });
