// @spartan-ng/ui-collapsible-brain/esm2022/lib/brn-collapsible-content.component.mjs
import { Component, ElementRef, Input, computed, inject, signal } from '@angular/core';
import { BrnCollapsibleComponent } from './brn-collapsible.component';
import * as i0 from "@angular/core";
export class BrnCollapsibleContentComponent {
    set id(value) {
        if (!value || !this._collapsible)
            return;
        this._collapsible.contentId.set(value);
    }
    constructor() {
        this._collapsible = inject(BrnCollapsibleComponent, { optional: true });
        this._elementRef = inject(ElementRef);
        this._height = signal(0);
        this._width = signal(0);
        this.contentId = this._collapsible?.contentId ?? signal(undefined).asReadonly;
        this.state = this._collapsible?.state ?? signal('closed').asReadonly();
        this.computedStyles = computed(() => {
            const height = this._height();
            const width = this._width();
            return {
                [`--radix-collapsible-content-height`]: height ? `${height}px` : undefined,
                [`--radix-collapsible-content-width`]: width ? `${width}px` : undefined,
            };
        });
        if (!this._collapsible) {
            throw Error('Collapsible trigger directive can only be used inside a brn-collapsible element.');
        }
    }
    ngAfterContentInit() {
        if (typeof this._elementRef.nativeElement.getBoundingClientRect !== 'function')
            return;
        const rect = this._elementRef.nativeElement.getBoundingClientRect();
        if (!rect)
            return;
        this._width.set(rect.width);
        this._height.set(rect.height);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnCollapsibleContentComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.3.5", type: BrnCollapsibleContentComponent, isStandalone: true, selector: "brn-collapsible-content", inputs: { id: "id" }, host: { properties: { "hidden": "state() === \"closed\"", "attr.data-state": "state()", "style": "computedStyles()", "id": "contentId()" } }, ngImport: i0, template: `
		<ng-content />
	`, isInline: true }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnCollapsibleContentComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'brn-collapsible-content',
                    standalone: true,
                    host: {
                        '[hidden]': 'state() === "closed"',
                        '[attr.data-state]': 'state()',
                        '[style]': 'computedStyles()',
                        '[id]': 'contentId()',
                    },
                    template: `
		<ng-content />
	`,
                }]
        }], ctorParameters: () => [], propDecorators: { id: [{
                type: Input
            }] } });
