// @spartan-ng/ui-collapsible-brain/esm2022/lib/brn-collapsible-trigger.directive.mjs
import { Directive, inject, signal } from '@angular/core';
import { BrnCollapsibleComponent } from './brn-collapsible.component';
import * as i0 from "@angular/core";
export class BrnCollapsibleTriggerDirective {
    constructor() {
        this._collapsible = inject(BrnCollapsibleComponent, { optional: true });
        this.state = this._collapsible?.state ?? signal(false).asReadonly();
        this.disabled = this._collapsible?.collapsibleDisabled ?? signal(undefined).asReadonly();
        this.ariaControls = this._collapsible?.contentId;
        if (!this._collapsible) {
            throw Error('Collapsible trigger directive can only be used inside a brn-collapsible element.');
        }
    }
    toggleCollapsible() {
        this._collapsible?.toggle();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnCollapsibleTriggerDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.5", type: BrnCollapsibleTriggerDirective, isStandalone: true, selector: "button[brnCollapsibleTrigger]", host: { listeners: { "click": "toggleCollapsible()" }, properties: { "attr.data-state": "state()", "attr.disabled": "disabled()", "attr.aria-expanded": "state() === \"open\"", "attr.aria-controls": "ariaControls()" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnCollapsibleTriggerDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'button[brnCollapsibleTrigger]',
                    standalone: true,
                    host: {
                        '[attr.data-state]': 'state()',
                        '[attr.disabled]': 'disabled()',
                        '[attr.aria-expanded]': 'state() === "open"',
                        '[attr.aria-controls]': 'ariaControls()',
                        '(click)': 'toggleCollapsible()',
                    },
                }]
        }], ctorParameters: () => [] });
