// @spartan-ng/ui-checkbox-brain/esm2022/index.mjs
import { NgModule } from '@angular/core';
import { BrnCheckboxComponent } from './lib/brn-checkbox.component';
import * as i0 from "@angular/core";
export * from './lib/brn-checkbox.component';
export const BrnCheckboxImports = [BrnCheckboxComponent];
export class BrnCheckboxModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnCheckboxModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.5", ngImport: i0, type: BrnCheckboxModule, imports: [BrnCheckboxComponent], exports: [BrnCheckboxComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnCheckboxModule }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnCheckboxModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...BrnCheckboxImports],
                    exports: [...BrnCheckboxImports],
                }]
        }] });
