// @spartan-ng/ui-command-brain/esm2022/index.mjs
import { NgModule } from '@angular/core';
import { CmdkModule } from '@ngxpert/cmdk';
import { BrnCommandEmptyDirective } from './lib/brn-command-empty.directive';
import { BrnCommandGroupComponent } from './lib/brn-command-group.component';
import { BrnCommandInputDirective } from './lib/brn-command-input.directive';
import { BrnCommandItemDirective } from './lib/brn-command-item.directive';
import { BrnCommandListComponent } from './lib/brn-command-list.component';
import { BrnCommandLoaderDirective } from './lib/brn-command-loader.directive';
import { BrnCommandSeparatorComponent } from './lib/brn-command-separator.component';
import { BrnCommandComponent } from './lib/brn-command.component';
import * as i0 from "@angular/core";
export * from './lib/brn-command-empty.directive';
export * from './lib/brn-command-group.component';
export * from './lib/brn-command-input.directive';
export * from './lib/brn-command-item.directive';
export * from './lib/brn-command-list.component';
export * from './lib/brn-command-loader.directive';
export * from './lib/brn-command-separator.component';
export * from './lib/brn-command.component';
export const BrnCommandImports = [
    CmdkModule,
    BrnCommandComponent,
    BrnCommandEmptyDirective,
    BrnCommandGroupComponent,
    BrnCommandInputDirective,
    BrnCommandItemDirective,
    BrnCommandListComponent,
    BrnCommandLoaderDirective,
    BrnCommandSeparatorComponent,
    BrnCommandSeparatorComponent,
];
export class BrnCommandModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnCommandModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.5", ngImport: i0, type: BrnCommandModule, imports: [CmdkModule,
            BrnCommandComponent,
            BrnCommandEmptyDirective,
            BrnCommandGroupComponent,
            BrnCommandInputDirective,
            BrnCommandItemDirective,
            BrnCommandListComponent,
            BrnCommandLoaderDirective,
            BrnCommandSeparatorComponent,
            BrnCommandSeparatorComponent], exports: [CmdkModule,
            BrnCommandComponent,
            BrnCommandEmptyDirective,
            BrnCommandGroupComponent,
            BrnCommandInputDirective,
            BrnCommandItemDirective,
            BrnCommandListComponent,
            BrnCommandLoaderDirective,
            BrnCommandSeparatorComponent,
            BrnCommandSeparatorComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnCommandModule, imports: [CmdkModule,
            BrnCommandGroupComponent,
            CmdkModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnCommandModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...BrnCommandImports],
                    exports: [...BrnCommandImports],
                }]
        }] });
