// @spartan-ng/ui-avatar-brain/esm2022/lib/util/initials.pipe.mjs
import { Pipe } from '@angular/core';
import * as i0 from "@angular/core";
const toInitial = (capitalize = true) => (word) => {
    const initial = word.charAt(0);
    return capitalize ? initial.toLocaleUpperCase() : initial;
};
const firstAndLast = (initials) => `${initials[0]}${initials[initials.length - 1]}`;
export class InitialsPipe {
    transform(name, capitalize = true, firstAndLastOnly = true, delimiter = ' ') {
        if (!name)
            return '';
        const initials = name.trim().split(delimiter).filter(Boolean).map(toInitial(capitalize));
        if (firstAndLastOnly && initials.length > 1)
            return firstAndLast(initials);
        return initials.join('');
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: InitialsPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe }); }
    static { this.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "17.3.5", ngImport: i0, type: InitialsPipe, isStandalone: true, name: "initials" }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: InitialsPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'initials',
                    standalone: true,
                }]
        }] });
