// @spartan-ng/ui-table-brain/esm2022/lib/brn-cell-def.directive.mjs
import { CdkCellDef } from '@angular/cdk/table';
// biome-ignore lint/style/useImportType: <explanation>
import { Directive, TemplateRef } from '@angular/core';
import * as i0 from "@angular/core";
export class BrnCellDefDirective extends CdkCellDef {
    constructor(template) {
        super(template);
        this.template = template;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnCellDefDirective, deps: [{ token: i0.TemplateRef }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.5", type: BrnCellDefDirective, isStandalone: true, selector: "[brnCellDef]", exportAs: ["brnCellDef"], usesInheritance: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnCellDefDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[brnCellDef]',
                    exportAs: 'brnCellDef',
                }]
        }], ctorParameters: () => [{ type: i0.TemplateRef }] });
