// @spartan-ng/ui-table-brain/esm2022/lib/brn-column-def.component.mjs
import { CdkColumnDef, CdkTableModule, } from '@angular/cdk/table';
import { ChangeDetectionStrategy, Component, ContentChild, Input, ViewChild, ViewEncapsulation, input, } from '@angular/core';
import { BrnCellDefDirective } from './brn-cell-def.directive';
import { BrnFooterDefDirective } from './brn-footer-def.directive';
import { BrnHeaderDefDirective } from './brn-header-def.directive';
import * as i0 from "@angular/core";
import * as i1 from "@angular/cdk/table";
export class BrnColumnDefComponent {
    constructor() {
        this._name = '';
        this.class = input('');
    }
    get columnDef() {
        return this._columnDef;
    }
    get cell() {
        return this._columnDef.cell;
    }
    get name() {
        return this._name;
    }
    set name(value) {
        this._name = value;
        if (!this._columnDef)
            return;
        this._columnDef.name = value;
    }
    ngAfterContentChecked() {
        this._columnDef.name = this.name;
        if (this._cellDef) {
            this._columnDef.cell = this._cellDef;
        }
        if (this._headerCellDef) {
            this._columnDef.headerCell = this._headerCellDef;
        }
        if (this._footerCellDef) {
            this._columnDef.footerCell = this._footerCellDef;
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnColumnDefComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.1.0", version: "17.3.5", type: BrnColumnDefComponent, isStandalone: true, selector: "brn-column-def", inputs: { name: { classPropertyName: "name", publicName: "name", isSignal: false, isRequired: false, transformFunction: null }, class: { classPropertyName: "class", publicName: "class", isSignal: true, isRequired: false, transformFunction: null } }, queries: [{ propertyName: "_cellDef", first: true, predicate: BrnCellDefDirective, descendants: true, static: true }, { propertyName: "_footerCellDef", first: true, predicate: BrnFooterDefDirective, descendants: true, static: true }, { propertyName: "_headerCellDef", first: true, predicate: BrnHeaderDefDirective, descendants: true, static: true }], viewQueries: [{ propertyName: "_columnDef", first: true, predicate: CdkColumnDef, descendants: true, static: true }], ngImport: i0, template: `
		<ng-container [cdkColumnDef]="name">
			<ng-content select="[brnHeaderDef]" />
			<ng-content select="[brnCellDef]" />
			<ng-content select="[brnFooterDef]" />
		</ng-container>
	`, isInline: true, dependencies: [{ kind: "ngmodule", type: CdkTableModule }, { kind: "directive", type: i1.CdkColumnDef, selector: "[cdkColumnDef]", inputs: ["cdkColumnDef", "sticky", "stickyEnd"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnColumnDefComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'brn-column-def',
                    standalone: true,
                    imports: [CdkTableModule],
                    template: `
		<ng-container [cdkColumnDef]="name">
			<ng-content select="[brnHeaderDef]" />
			<ng-content select="[brnCellDef]" />
			<ng-content select="[brnFooterDef]" />
		</ng-container>
	`,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    encapsulation: ViewEncapsulation.None,
                }]
        }], propDecorators: { name: [{
                type: Input
            }], _columnDef: [{
                type: ViewChild,
                args: [CdkColumnDef, { static: true }]
            }], _cellDef: [{
                type: ContentChild,
                args: [BrnCellDefDirective, { static: true }]
            }], _footerCellDef: [{
                type: ContentChild,
                args: [BrnFooterDefDirective, { static: true }]
            }], _headerCellDef: [{
                type: ContentChild,
                args: [BrnHeaderDefDirective, { static: true }]
            }] } });
