// @spartan-ng/ui-table-brain/esm2022/lib/brn-footer-def.directive.mjs
import { CdkFooterCellDef } from '@angular/cdk/table';
// biome-ignore lint/style/useImportType: <explanation>
import { Directive, TemplateRef } from '@angular/core';
import * as i0 from "@angular/core";
export class BrnFooterDefDirective extends CdkFooterCellDef {
    constructor(template) {
        super(template);
        this.template = template;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnFooterDefDirective, deps: [{ token: i0.TemplateRef }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.5", type: BrnFooterDefDirective, isStandalone: true, selector: "[brnFooterDef]", exportAs: ["brnFooterDef"], usesInheritance: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnFooterDefDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[brnFooterDef]',
                    exportAs: 'brnFooterDef',
                }]
        }], ctorParameters: () => [{ type: i0.TemplateRef }] });
