// @spartan-ng/ui-command-brain/esm2022/lib/brn-command.component.mjs
import { Component } from '@angular/core';
import { CmdkService, CommandComponent } from '@ngxpert/cmdk';
import * as i0 from "@angular/core";
export class BrnCommandComponent extends CommandComponent {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnCommandComponent, deps: null, target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.3.5", type: BrnCommandComponent, isStandalone: true, selector: "brn-cmd", providers: [CmdkService], usesInheritance: true, ngImport: i0, template: `
		<ng-content />
	`, isInline: true }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnCommandComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'brn-cmd',
                    standalone: true,
                    providers: [CmdkService],
                    template: `
		<ng-content />
	`,
                }]
        }] });
