// @spartan-ng/ui-command-brain/esm2022/lib/brn-command-item.directive.mjs
import { __decorate } from "tslib";
import { Directive, forwardRef } from '@angular/core';
import { UntilDestroy } from '@ngneat/until-destroy';
import { ItemDirective } from '@ngxpert/cmdk';
import * as i0 from "@angular/core";
let BrnCommandItemDirective = class BrnCommandItemDirective extends ItemDirective {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnCommandItemDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.5", type: BrnCommandItemDirective, isStandalone: true, selector: "[brnCmdItem]", host: { attributes: { "tabindex": "-1" }, classAttribute: "cmdk-item" }, providers: [
            {
                provide: ItemDirective,
                useExisting: forwardRef(() => BrnCommandItemDirective),
            },
        ], usesInheritance: true, ngImport: i0 }); }
};
BrnCommandItemDirective = __decorate([
    UntilDestroy()
], BrnCommandItemDirective);
export { BrnCommandItemDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnCommandItemDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[brnCmdItem]',
                    standalone: true,
                    providers: [
                        {
                            provide: ItemDirective,
                            useExisting: forwardRef(() => BrnCommandItemDirective),
                        },
                    ],
                    host: {
                        tabindex: '-1',
                        class: 'cmdk-item',
                    },
                }]
        }] });
