// @spartan-ng/ui-command-brain/esm2022/lib/brn-command-group.component.mjs
import { NgIf } from '@angular/common';
import { Component, forwardRef } from '@angular/core';
import { DynamicViewModule } from '@ngneat/overview';
import { GroupComponent } from '@ngxpert/cmdk';
import * as i0 from "@angular/core";
import * as i1 from "@ngneat/overview";
export class BrnCommandGroupComponent extends GroupComponent {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnCommandGroupComponent, deps: null, target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.3.5", type: BrnCommandGroupComponent, isStandalone: true, selector: "brn-cmd-group", providers: [
            {
                provide: GroupComponent,
                useExisting: forwardRef(() => BrnCommandGroupComponent),
            },
        ], usesInheritance: true, ngImport: i0, template: `
		<div role="presentation" *ngIf="label" class="cmdk-group-label">
			<ng-container *dynamicView="label"></ng-container>
		</div>
		<div class="cmdk-group-content" role="group" [attr.aria-label]="ariaLabel">
			<ng-content></ng-content>
		</div>
	`, isInline: true, dependencies: [{ kind: "directive", type: NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { kind: "ngmodule", type: DynamicViewModule }, { kind: "directive", type: i1.DynamicViewDirective, selector: "[dynamicView]", inputs: ["dynamicView", "dynamicViewInjector", "dynamicViewContext", "dynamicViewInputs"] }] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnCommandGroupComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'brn-cmd-group',
                    standalone: true,
                    imports: [NgIf, DynamicViewModule],
                    providers: [
                        {
                            provide: GroupComponent,
                            useExisting: forwardRef(() => BrnCommandGroupComponent),
                        },
                    ],
                    template: `
		<div role="presentation" *ngIf="label" class="cmdk-group-label">
			<ng-container *dynamicView="label"></ng-container>
		</div>
		<div class="cmdk-group-content" role="group" [attr.aria-label]="ariaLabel">
			<ng-content></ng-content>
		</div>
	`,
                }]
        }] });
