// @spartan-ng/ui-command-brain/esm2022/lib/brn-command-loader.directive.mjs
import { Directive, forwardRef } from '@angular/core';
import { LoaderDirective } from '@ngxpert/cmdk';
import * as i0 from "@angular/core";
export class BrnCommandLoaderDirective extends LoaderDirective {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnCommandLoaderDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.5", type: BrnCommandLoaderDirective, isStandalone: true, selector: "[brnCmdLoader]", host: { classAttribute: "cmdk-loader" }, providers: [
            {
                provide: LoaderDirective,
                useExisting: forwardRef(() => BrnCommandLoaderDirective),
            },
        ], usesInheritance: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnCommandLoaderDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[brnCmdLoader]',
                    standalone: true,
                    providers: [
                        {
                            provide: LoaderDirective,
                            useExisting: forwardRef(() => BrnCommandLoaderDirective),
                        },
                    ],
                    host: {
                        class: 'cmdk-loader',
                    },
                }]
        }] });
