// @spartan-ng/ui-command-brain/esm2022/lib/brn-command-input.directive.mjs
import { Directive, forwardRef } from '@angular/core';
import { InputDirective } from '@ngxpert/cmdk';
import * as i0 from "@angular/core";
export class BrnCommandInputDirective extends InputDirective {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnCommandInputDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.5", type: BrnCommandInputDirective, isStandalone: true, selector: "input[brnCmdInput]", host: { classAttribute: "cmdk-input" }, providers: [
            {
                provide: InputDirective,
                useExisting: forwardRef(() => BrnCommandInputDirective),
            },
        ], usesInheritance: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnCommandInputDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'input[brnCmdInput]',
                    standalone: true,
                    providers: [
                        {
                            provide: InputDirective,
                            useExisting: forwardRef(() => BrnCommandInputDirective),
                        },
                    ],
                    host: {
                        class: 'cmdk-input',
                    },
                }]
        }] });
