// @spartan-ng/ui-command-brain/esm2022/lib/brn-command-separator.component.mjs
import { Component, forwardRef } from '@angular/core';
import { SeparatorComponent } from '@ngxpert/cmdk';
import * as i0 from "@angular/core";
export class BrnCommandSeparatorComponent extends SeparatorComponent {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnCommandSeparatorComponent, deps: null, target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.3.5", type: BrnCommandSeparatorComponent, isStandalone: true, selector: "brn-cmd-separator", host: { classAttribute: "cmdk-separator" }, providers: [
            {
                provide: SeparatorComponent,
                useExisting: forwardRef(() => BrnCommandSeparatorComponent),
            },
        ], usesInheritance: true, ngImport: i0, template: '<hr>', isInline: true }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnCommandSeparatorComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'brn-cmd-separator',
                    standalone: true,
                    template: '<hr>',
                    providers: [
                        {
                            provide: SeparatorComponent,
                            useExisting: forwardRef(() => BrnCommandSeparatorComponent),
                        },
                    ],
                    host: {
                        class: 'cmdk-separator',
                    },
                }]
        }] });
