// @spartan-ng/ui-command-brain/esm2022/lib/brn-command-list.component.mjs
import { Component, forwardRef } from '@angular/core';
import { ListComponent } from '@ngxpert/cmdk';
import * as i0 from "@angular/core";
export class BrnCommandListComponent extends ListComponent {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnCommandListComponent, deps: null, target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.3.5", type: BrnCommandListComponent, isStandalone: true, selector: "brn-cmd-list", providers: [
            {
                provide: ListComponent,
                useExisting: forwardRef(() => BrnCommandListComponent),
            },
        ], usesInheritance: true, ngImport: i0, template: `
		<div class="cmdk-list-content" role="listbox" [attr.aria-label]="ariaLabel" #height>
			<ng-content></ng-content>
		</div>
	`, isInline: true }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnCommandListComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'brn-cmd-list',
                    standalone: true,
                    providers: [
                        {
                            provide: ListComponent,
                            useExisting: forwardRef(() => BrnCommandListComponent),
                        },
                    ],
                    template: `
		<div class="cmdk-list-content" role="listbox" [attr.aria-label]="ariaLabel" #height>
			<ng-content></ng-content>
		</div>
	`,
                }]
        }] });
