// @spartan-ng/ui-command-brain/esm2022/lib/brn-command-empty.directive.mjs
import { Directive, forwardRef } from '@angular/core';
import { EmptyDirective } from '@ngxpert/cmdk';
import * as i0 from "@angular/core";
export class BrnCommandEmptyDirective extends EmptyDirective {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnCommandEmptyDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.5", type: BrnCommandEmptyDirective, isStandalone: true, selector: "[brnCmdEmpty]", host: { classAttribute: "cmdk-empty" }, providers: [
            {
                provide: EmptyDirective,
                useExisting: forwardRef(() => BrnCommandEmptyDirective),
            },
        ], usesInheritance: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnCommandEmptyDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[brnCmdEmpty]',
                    standalone: true,
                    providers: [
                        {
                            provide: EmptyDirective,
                            useExisting: forwardRef(() => BrnCommandEmptyDirective),
                        },
                    ],
                    host: {
                        class: 'cmdk-empty',
                    },
                }]
        }] });
