// @spartan-ng/ui-radiogroup-brain/esm2022/index.mjs
import { NgModule } from '@angular/core';
import { BrnRadioGroupComponent } from './lib/brn-radio-group.component';
import { BrnRadioComponent } from './lib/brn-radio.component';
import * as i0 from "@angular/core";
export * from './lib/brn-radio-change';
export * from './lib/brn-radio-group.component';
export * from './lib/brn-radio.component';
export const BrnRadioGroupImports = [BrnRadioGroupComponent, BrnRadioComponent];
export class BrnRadioGroupModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnRadioGroupModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.5", ngImport: i0, type: BrnRadioGroupModule, imports: [BrnRadioGroupComponent, BrnRadioComponent], exports: [BrnRadioGroupComponent, BrnRadioComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnRadioGroupModule }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnRadioGroupModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...BrnRadioGroupImports],
                    exports: [...BrnRadioGroupImports],
                }]
        }] });
