// @spartan-ng/ui-sheet-brain/esm2022/lib/brn-sheet.component.mjs
import { ChangeDetectionStrategy, Component, Input, ViewEncapsulation, forwardRef, signal } from '@angular/core';
import { BrnDialogComponent } from '@spartan-ng/ui-dialog-brain';
import * as i0 from "@angular/core";
export class BrnSheetComponent extends BrnDialogComponent {
    constructor() {
        super(...arguments);
        this._side = signal('top');
        this.side = this._side.asReadonly();
    }
    /* eslint-disable-next-line @angular-eslint/no-input-rename */
    set setSide(side) {
        this._side.set(side);
        if (side === 'top') {
            this.positionStrategy = this.positionBuilder.global().top();
        }
        if (side === 'bottom') {
            this.positionStrategy = this.positionBuilder.global().bottom();
        }
        if (side === 'left') {
            this.positionStrategy = this.positionBuilder.global().left();
        }
        if (side === 'right') {
            this.positionStrategy = this.positionBuilder.global().right();
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnSheetComponent, deps: null, target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.3.5", type: BrnSheetComponent, isStandalone: true, selector: "brn-sheet", inputs: { setSide: ["side", "setSide"] }, providers: [
            {
                provide: BrnDialogComponent,
                useExisting: forwardRef(() => BrnSheetComponent),
            },
        ], exportAs: ["brnSheet"], usesInheritance: true, ngImport: i0, template: `
		<ng-content />
	`, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnSheetComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'brn-sheet',
                    standalone: true,
                    template: `
		<ng-content />
	`,
                    providers: [
                        {
                            provide: BrnDialogComponent,
                            useExisting: forwardRef(() => BrnSheetComponent),
                        },
                    ],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    encapsulation: ViewEncapsulation.None,
                    exportAs: 'brnSheet',
                }]
        }], propDecorators: { setSide: [{
                type: Input,
                args: ['side']
            }] } });
