// @spartan-ng/ui-sheet-brain/esm2022/lib/brn-sheet-trigger.directive.mjs
import { Directive, Input, inject } from '@angular/core';
import { BrnDialogTriggerDirective } from '@spartan-ng/ui-dialog-brain';
import { BrnSheetComponent } from './brn-sheet.component';
import * as i0 from "@angular/core";
export class BrnSheetTriggerDirective extends BrnDialogTriggerDirective {
    constructor() {
        super(...arguments);
        this._sheet = inject(BrnSheetComponent, { optional: true });
    }
    open() {
        if (this._sheet && this.side) {
            this._sheet.setSide = this.side;
        }
        super.open();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnSheetTriggerDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.5", type: BrnSheetTriggerDirective, isStandalone: true, selector: "button[brnSheetTrigger]", inputs: { side: "side" }, usesInheritance: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnSheetTriggerDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'button[brnSheetTrigger]',
                    standalone: true,
                }]
        }], propDecorators: { side: [{
                type: Input
            }] } });
