// @spartan-ng/ui-sheet-brain/esm2022/lib/brn-sheet-content.directive.mjs
import { Directive, inject } from '@angular/core';
import { provideExposedSideProviderExisting, provideExposesStateProviderExisting, } from '@spartan-ng/ui-core';
import { BrnDialogContentDirective } from '@spartan-ng/ui-dialog-brain';
import { BrnSheetComponent } from './brn-sheet.component';
import * as i0 from "@angular/core";
export class BrnSheetContentDirective extends BrnDialogContentDirective {
    constructor() {
        super(...arguments);
        this.side = inject(BrnSheetComponent).side;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnSheetContentDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.5", type: BrnSheetContentDirective, isStandalone: true, selector: "[brnSheetContent]", providers: [
            provideExposesStateProviderExisting(() => BrnSheetContentDirective),
            provideExposedSideProviderExisting(() => BrnSheetContentDirective),
        ], usesInheritance: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnSheetContentDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[brnSheetContent]',
                    standalone: true,
                    providers: [
                        provideExposesStateProviderExisting(() => BrnSheetContentDirective),
                        provideExposedSideProviderExisting(() => BrnSheetContentDirective),
                    ],
                }]
        }] });
