// @spartan-ng/ui-sheet-brain/esm2022/lib/brn-sheet-overlay.component.mjs
import { ChangeDetectionStrategy, Component, ViewEncapsulation } from '@angular/core';
import { provideCustomClassSettableExisting } from '@spartan-ng/ui-core';
import { BrnDialogOverlayComponent } from '@spartan-ng/ui-dialog-brain';
import * as i0 from "@angular/core";
export class BrnSheetOverlayComponent extends BrnDialogOverlayComponent {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnSheetOverlayComponent, deps: null, target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.3.5", type: BrnSheetOverlayComponent, isStandalone: true, selector: "brn-sheet-overlay", providers: [provideCustomClassSettableExisting(() => BrnSheetOverlayComponent)], usesInheritance: true, ngImport: i0, template: ``, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnSheetOverlayComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'brn-sheet-overlay',
                    standalone: true,
                    providers: [provideCustomClassSettableExisting(() => BrnSheetOverlayComponent)],
                    template: ``,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    encapsulation: ViewEncapsulation.None,
                }]
        }] });
