// @spartan-ng/ui-progress-brain/esm2022/index.mjs
import { NgModule } from '@angular/core';
import { BrnProgressIndicatorComponent } from './lib/brn-progress-indicator.component';
import { BrnProgressComponent } from './lib/brn-progress.component';
import * as i0 from "@angular/core";
export * from './lib/brn-progress-indicator.component';
export * from './lib/brn-progress.component';
export const BrnProgressImports = [BrnProgressComponent, BrnProgressIndicatorComponent];
export class BrnProgressModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnProgressModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.5", ngImport: i0, type: BrnProgressModule, imports: [BrnProgressComponent, BrnProgressIndicatorComponent], exports: [BrnProgressComponent, BrnProgressIndicatorComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnProgressModule }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnProgressModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...BrnProgressImports],
                    exports: [...BrnProgressImports],
                }]
        }] });
