// @spartan-ng/ui-avatar-brain/esm2022/lib/fallback/brn-avatar-fallback.directive.mjs
import { Directive, ElementRef, Input, booleanAttribute, inject, signal } from '@angular/core';
import * as i0 from "@angular/core";
export class BrnAvatarFallbackDirective {
    constructor() {
        this.element = inject(ElementRef).nativeElement;
        this.userCls = signal('');
        this.useAutoColor = signal(false);
        this.textContent = inject(ElementRef).nativeElement.textContent;
    }
    getTextContent() {
        return this.element.textContent;
    }
    set class(cls) {
        this.userCls.set(cls);
    }
    set autoColor(value) {
        this.useAutoColor.set(value);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnAvatarFallbackDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "16.1.0", version: "17.3.5", type: BrnAvatarFallbackDirective, isStandalone: true, selector: "[brnAvatarFallback]", inputs: { class: "class", autoColor: ["autoColor", "autoColor", booleanAttribute] }, exportAs: ["avatarFallback"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnAvatarFallbackDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[brnAvatarFallback]',
                    standalone: true,
                    exportAs: 'avatarFallback',
                }]
        }], propDecorators: { class: [{
                type: Input
            }], autoColor: [{
                type: Input,
                args: [{ transform: booleanAttribute }]
            }] } });
