// @spartan-ng/ui-table-brain/esm2022/index.mjs
import { NgModule } from '@angular/core';
import { BrnCellDefDirective } from './lib/brn-cell-def.directive';
import { BrnColumnDefComponent } from './lib/brn-column-def.component';
import { BrnFooterDefDirective } from './lib/brn-footer-def.directive';
import { BrnHeaderDefDirective } from './lib/brn-header-def.directive';
import { BrnPaginatorDirective } from './lib/brn-paginator.directive';
import { BrnTableComponent } from './lib/brn-table.component';
import * as i0 from "@angular/core";
export { BrnCellDefDirective } from './lib/brn-cell-def.directive';
export { BrnColumnDefComponent } from './lib/brn-column-def.component';
export { BrnColumnManager, useBrnColumnManager } from './lib/brn-column-manager';
export { BrnFooterDefDirective } from './lib/brn-footer-def.directive';
export { BrnHeaderDefDirective } from './lib/brn-header-def.directive';
export { BrnPaginatorDirective } from './lib/brn-paginator.directive';
export { BrnTableComponent } from './lib/brn-table.component';
export const BrnTableImports = [
    BrnCellDefDirective,
    BrnColumnDefComponent,
    BrnFooterDefDirective,
    BrnHeaderDefDirective,
    BrnTableComponent,
    BrnPaginatorDirective,
];
export class BrnTableModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnTableModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.5", ngImport: i0, type: BrnTableModule, imports: [BrnCellDefDirective,
            BrnColumnDefComponent,
            BrnFooterDefDirective,
            BrnHeaderDefDirective,
            BrnTableComponent,
            BrnPaginatorDirective], exports: [BrnCellDefDirective,
            BrnColumnDefComponent,
            BrnFooterDefDirective,
            BrnHeaderDefDirective,
            BrnTableComponent,
            BrnPaginatorDirective] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnTableModule, imports: [BrnColumnDefComponent,
            BrnTableComponent] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnTableModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...BrnTableImports],
                    exports: [...BrnTableImports],
                }]
        }] });
