// @spartan-ng/ui-accordion-brain/esm2022/lib/brn-accordion-trigger.directive.mjs
import { Directive, ElementRef, inject } from '@angular/core';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import { rxHostPressedListener } from '@spartan-ng/ui-core';
import { fromEvent } from 'rxjs';
import { BrnAccordionItemDirective } from './brn-accordion-item.directive';
import { BrnAccordionDirective } from './brn-accordion.directive';
import * as i0 from "@angular/core";
export class BrnAccordionTriggerDirective {
    constructor() {
        this._accordion = inject(BrnAccordionDirective);
        this._item = inject(BrnAccordionItemDirective);
        this._elementRef = inject(ElementRef);
        this._hostPressedListener = rxHostPressedListener();
        this.state = this._item.state;
        this.id = 'brn-accordion-trigger-' + this._item.id;
        this.ariaControls = 'brn-accordion-content-' + this._item.id;
        if (!this._accordion) {
            throw Error('Accordion trigger can only be used inside an Accordion. Add brnAccordion to ancestor.');
        }
        if (!this._item) {
            throw Error('Accordion trigger can only be used inside an AccordionItem. Add brnAccordionItem to parent.');
        }
        this._hostPressedListener.subscribe(() => {
            this._accordion.toggleItem(this._item.id);
        });
        fromEvent(this._elementRef.nativeElement, 'focus')
            .pipe(takeUntilDestroyed())
            .subscribe(() => {
            this._accordion.setActiveItem(this);
        });
    }
    focus() {
        this._elementRef.nativeElement.focus();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnAccordionTriggerDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.5", type: BrnAccordionTriggerDirective, isStandalone: true, selector: "[brnAccordionTrigger]", host: { attributes: { "role": "heading", "aria-level": "3" }, properties: { "attr.data-state": "state()", "attr.aria-expanded": "state() === \"open\"", "attr.aria-controls": "ariaControls", "id": "id" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnAccordionTriggerDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[brnAccordionTrigger]',
                    standalone: true,
                    host: {
                        '[attr.data-state]': 'state()',
                        '[attr.aria-expanded]': 'state() === "open"',
                        '[attr.aria-controls]': 'ariaControls',
                        role: 'heading',
                        'aria-level': '3',
                        '[id]': 'id',
                    },
                }]
        }], ctorParameters: () => [] });
