// @spartan-ng/ui-accordion-brain/esm2022/lib/brn-accordion-item.directive.mjs
import { coerceBooleanProperty } from '@angular/cdk/coercion';
import { Directive, computed, effect, inject, input, untracked } from '@angular/core';
import { BrnAccordionDirective } from './brn-accordion.directive';
import * as i0 from "@angular/core";
let itemIdGenerator = 0;
export class BrnAccordionItemDirective {
    constructor() {
        this._accordion = inject(BrnAccordionDirective);
        this.isOpened = input(false, { transform: coerceBooleanProperty });
        this.id = itemIdGenerator++;
        this.state = computed(() => (this._accordion.openItemIds().includes(this.id) ? 'open' : 'closed'));
        if (!this._accordion) {
            throw Error('Accordion trigger can only be used inside an Accordion. Add brnAccordion to ancestor.');
        }
        effect(() => {
            const isOpened = this.isOpened();
            untracked(() => {
                if (isOpened) {
                    this._accordion.openItem(this.id);
                }
                else {
                    this._accordion.closeItem(this.id);
                }
            });
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnAccordionItemDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "17.3.5", type: BrnAccordionItemDirective, isStandalone: true, selector: "[brnAccordionItem]", inputs: { isOpened: { classPropertyName: "isOpened", publicName: "isOpened", isSignal: true, isRequired: false, transformFunction: null } }, host: { properties: { "attr.data-state": "state()" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnAccordionItemDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[brnAccordionItem]',
                    standalone: true,
                    host: {
                        '[attr.data-state]': 'state()',
                    },
                }]
        }], ctorParameters: () => [] });
