// @ngxpert/cmdk/esm2022/lib/directives/empty/empty.directive.mjs
import { Directive, inject, Renderer2, TemplateRef, ViewContainerRef, } from '@angular/core';
import * as i0 from "@angular/core";
export class EmptyDirective {
    constructor() {
        this._hasView = false;
        this._templateRef = inject((TemplateRef));
        this._viewContainer = inject(ViewContainerRef);
        this._renderer2 = inject(Renderer2);
    }
    set cmdkEmpty(condition) {
        if (condition && !this._hasView) {
            const emb = this._viewContainer.createEmbeddedView(this._templateRef);
            this._renderer2.addClass(emb.rootNodes[0], 'cmdk-empty');
            this._hasView = true;
        }
        else if (!condition && this._hasView) {
            this._viewContainer.clear();
            this._hasView = false;
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: EmptyDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: EmptyDirective, isStandalone: true, selector: "[cmdkEmpty]", host: { classAttribute: "cmdk-empty" }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: EmptyDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[cmdkEmpty]',
                    // eslint-disable-next-line @angular-eslint/no-host-metadata-property
                    host: {
                        class: 'cmdk-empty',
                    },
                    standalone: true,
                }]
        }] });
