// @spartan-ng/ui-tabs-brain/esm2022/index.mjs
import { NgModule } from '@angular/core';
import { BrnTabsContentDirective } from './lib/brn-tabs-content.directive';
import { BrnTabsListDirective } from './lib/brn-tabs-list.directive';
import { BrnTabsTriggerDirective } from './lib/brn-tabs-trigger.directive';
import { BrnTabsDirective } from './lib/brn-tabs.directive';
import * as i0 from "@angular/core";
export * from './lib/brn-tabs-content.directive';
export * from './lib/brn-tabs-list.directive';
export * from './lib/brn-tabs-trigger.directive';
export * from './lib/brn-tabs.directive';
export const BrnTabsImports = [
    BrnTabsDirective,
    BrnTabsListDirective,
    BrnTabsTriggerDirective,
    BrnTabsContentDirective,
];
export class BrnTabsModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnTabsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.5", ngImport: i0, type: BrnTabsModule, imports: [BrnTabsDirective,
            BrnTabsListDirective,
            BrnTabsTriggerDirective,
            BrnTabsContentDirective], exports: [BrnTabsDirective,
            BrnTabsListDirective,
            BrnTabsTriggerDirective,
            BrnTabsContentDirective] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnTabsModule }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnTabsModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...BrnTabsImports],
                    exports: [...BrnTabsImports],
                }]
        }] });
