// @spartan-ng/ui-sheet-brain/esm2022/index.mjs
import { NgModule } from '@angular/core';
import { BrnSheetCloseDirective } from './lib/brn-sheet-close.directive';
import { BrnSheetContentDirective } from './lib/brn-sheet-content.directive';
import { BrnSheetDescriptionDirective } from './lib/brn-sheet-description.directive';
import { BrnSheetOverlayComponent } from './lib/brn-sheet-overlay.component';
import { BrnSheetTitleDirective } from './lib/brn-sheet-title.directive';
import { BrnSheetTriggerDirective } from './lib/brn-sheet-trigger.directive';
import { BrnSheetComponent } from './lib/brn-sheet.component';
import * as i0 from "@angular/core";
export * from './lib/brn-sheet-close.directive';
export * from './lib/brn-sheet-content.directive';
export * from './lib/brn-sheet-description.directive';
export * from './lib/brn-sheet-overlay.component';
export * from './lib/brn-sheet-title.directive';
export * from './lib/brn-sheet-trigger.directive';
export * from './lib/brn-sheet.component';
export const BrnSheetImports = [
    BrnSheetComponent,
    BrnSheetOverlayComponent,
    BrnSheetTriggerDirective,
    BrnSheetCloseDirective,
    BrnSheetContentDirective,
    BrnSheetTitleDirective,
    BrnSheetDescriptionDirective,
];
export class BrnSheetModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnSheetModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.5", ngImport: i0, type: BrnSheetModule, imports: [BrnSheetComponent,
            BrnSheetOverlayComponent,
            BrnSheetTriggerDirective,
            BrnSheetCloseDirective,
            BrnSheetContentDirective,
            BrnSheetTitleDirective,
            BrnSheetDescriptionDirective], exports: [BrnSheetComponent,
            BrnSheetOverlayComponent,
            BrnSheetTriggerDirective,
            BrnSheetCloseDirective,
            BrnSheetContentDirective,
            BrnSheetTitleDirective,
            BrnSheetDescriptionDirective] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnSheetModule }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.5", ngImport: i0, type: BrnSheetModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...BrnSheetImports],
                    exports: [...BrnSheetImports],
                }]
        }] });
